<!doctype html>
<html lang="en" dir="ltr" class="docs-wrapper docs-doc-page docs-version-current plugin-docs plugin-id-default docs-doc-id-常用分析流程及教程/复现">
<head>
<meta charset="UTF-8">
<meta name="generator" content="Docusaurus v2.4.1">
<title data-rh="true">文章复现 | My Site</title><meta data-rh="true" name="viewport" content="width=device-width,initial-scale=1"><meta data-rh="true" name="twitter:card" content="summary_large_image"><meta data-rh="true" property="og:image" content="https://your-docusaurus-test-site.com/BioinfoTech/img/docusaurus-social-card.jpg"><meta data-rh="true" name="twitter:image" content="https://your-docusaurus-test-site.com/BioinfoTech/img/docusaurus-social-card.jpg"><meta data-rh="true" property="og:url" content="https://your-docusaurus-test-site.com/BioinfoTech/常用分析流程及教程/复现"><meta data-rh="true" name="docusaurus_locale" content="en"><meta data-rh="true" name="docsearch:language" content="en"><meta data-rh="true" name="docusaurus_version" content="current"><meta data-rh="true" name="docusaurus_tag" content="docs-default-current"><meta data-rh="true" name="docsearch:version" content="current"><meta data-rh="true" name="docsearch:docusaurus_tag" content="docs-default-current"><meta data-rh="true" property="og:title" content="文章复现 | My Site"><meta data-rh="true" name="description" content="来自 @生信技能树"><meta data-rh="true" property="og:description" content="来自 @生信技能树"><link data-rh="true" rel="icon" href="/BioinfoTech/img/favicon.ico"><link data-rh="true" rel="canonical" href="https://your-docusaurus-test-site.com/BioinfoTech/常用分析流程及教程/复现"><link data-rh="true" rel="alternate" href="https://your-docusaurus-test-site.com/BioinfoTech/常用分析流程及教程/复现" hreflang="en"><link data-rh="true" rel="alternate" href="https://your-docusaurus-test-site.com/BioinfoTech/常用分析流程及教程/复现" hreflang="x-default"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.13.24/dist/katex.min.css" integrity="sha384-odtC+0UGzzFL/6PNoE8rX/SPcQDXBJ+uRepguP4QkPCm2LBxH3FA3y+fKSiJ+AmM" crossorigin="anonymous"><link rel="stylesheet" href="/BioinfoTech/assets/css/styles.a4d42eef.css">
<link rel="preload" href="/BioinfoTech/assets/js/runtime~main.9748e4a1.js" as="script">
<link rel="preload" href="/BioinfoTech/assets/js/main.3ef7b66f.js" as="script">
</head>
<body class="navigation-with-keyboard">
<script>!function(){function t(t){document.documentElement.setAttribute("data-theme",t)}var e=function(){var t=null;try{t=new URLSearchParams(window.location.search).get("docusaurus-theme")}catch(t){}return t}()||function(){var t=null;try{t=localStorage.getItem("theme")}catch(t){}return t}();t(null!==e?e:"light")}()</script><div id="__docusaurus">
<div role="region" aria-label="Skip to main content"><a class="skipToContent_fXgn" href="#__docusaurus_skipToContent_fallback">Skip to main content</a></div><nav aria-label="Main" class="navbar navbar--fixed-top"><div class="navbar__inner"><div class="navbar__items"><button aria-label="Toggle navigation bar" aria-expanded="false" class="navbar__toggle clean-btn" type="button"><svg width="30" height="30" viewBox="0 0 30 30" aria-hidden="true"><path stroke="currentColor" stroke-linecap="round" stroke-miterlimit="10" stroke-width="2" d="M4 7h22M4 15h22M4 23h22"></path></svg></button><a class="navbar__brand" href="/BioinfoTech/"><div class="navbar__logo"><img src="/BioinfoTech/img/logo.svg" alt="My Site Logo" class="themedImage_ToTc themedImage--light_HNdA"><img src="/BioinfoTech/img/logo.svg" alt="My Site Logo" class="themedImage_ToTc themedImage--dark_i4oU"></div><b class="navbar__title text--truncate">My Site</b></a><a aria-current="page" class="navbar__item navbar__link navbar__link--active" href="/BioinfoTech/intro">Tutorial</a></div><div class="navbar__items navbar__items--right"><a href="https://github.com/WhyLIM" target="_blank" rel="noopener noreferrer" class="navbar__item navbar__link">GitHub<svg width="13.5" height="13.5" aria-hidden="true" viewBox="0 0 24 24" class="iconExternalLink_nPIU"><path fill="currentColor" d="M21 13v10h-21v-19h12v2h-10v15h17v-8h2zm3-12h-10.988l4.035 4-6.977 7.07 2.828 2.828 6.977-7.07 4.125 4.172v-11z"></path></svg></a><div class="toggle_vylO colorModeToggle_DEke"><button class="clean-btn toggleButton_gllP toggleButtonDisabled_aARS" type="button" disabled="" title="Switch between dark and light mode (currently light mode)" aria-label="Switch between dark and light mode (currently light mode)" aria-live="polite"><svg viewBox="0 0 24 24" width="24" height="24" class="lightToggleIcon_pyhR"><path fill="currentColor" d="M12,9c1.65,0,3,1.35,3,3s-1.35,3-3,3s-3-1.35-3-3S10.35,9,12,9 M12,7c-2.76,0-5,2.24-5,5s2.24,5,5,5s5-2.24,5-5 S14.76,7,12,7L12,7z M2,13l2,0c0.55,0,1-0.45,1-1s-0.45-1-1-1l-2,0c-0.55,0-1,0.45-1,1S1.45,13,2,13z M20,13l2,0c0.55,0,1-0.45,1-1 s-0.45-1-1-1l-2,0c-0.55,0-1,0.45-1,1S19.45,13,20,13z M11,2v2c0,0.55,0.45,1,1,1s1-0.45,1-1V2c0-0.55-0.45-1-1-1S11,1.45,11,2z M11,20v2c0,0.55,0.45,1,1,1s1-0.45,1-1v-2c0-0.55-0.45-1-1-1C11.45,19,11,19.45,11,20z M5.99,4.58c-0.39-0.39-1.03-0.39-1.41,0 c-0.39,0.39-0.39,1.03,0,1.41l1.06,1.06c0.39,0.39,1.03,0.39,1.41,0s0.39-1.03,0-1.41L5.99,4.58z M18.36,16.95 c-0.39-0.39-1.03-0.39-1.41,0c-0.39,0.39-0.39,1.03,0,1.41l1.06,1.06c0.39,0.39,1.03,0.39,1.41,0c0.39-0.39,0.39-1.03,0-1.41 L18.36,16.95z M19.42,5.99c0.39-0.39,0.39-1.03,0-1.41c-0.39-0.39-1.03-0.39-1.41,0l-1.06,1.06c-0.39,0.39-0.39,1.03,0,1.41 s1.03,0.39,1.41,0L19.42,5.99z M7.05,18.36c0.39-0.39,0.39-1.03,0-1.41c-0.39-0.39-1.03-0.39-1.41,0l-1.06,1.06 c-0.39,0.39-0.39,1.03,0,1.41s1.03,0.39,1.41,0L7.05,18.36z"></path></svg><svg viewBox="0 0 24 24" width="24" height="24" class="darkToggleIcon_wfgR"><path fill="currentColor" d="M9.37,5.51C9.19,6.15,9.1,6.82,9.1,7.5c0,4.08,3.32,7.4,7.4,7.4c0.68,0,1.35-0.09,1.99-0.27C17.45,17.19,14.93,19,12,19 c-3.86,0-7-3.14-7-7C5,9.07,6.81,6.55,9.37,5.51z M12,3c-4.97,0-9,4.03-9,9s4.03,9,9,9s9-4.03,9-9c0-0.46-0.04-0.92-0.1-1.36 c-0.98,1.37-2.58,2.26-4.4,2.26c-2.98,0-5.4-2.42-5.4-5.4c0-1.81,0.89-3.42,2.26-4.4C12.92,3.04,12.46,3,12,3L12,3z"></path></svg></button></div><div class="searchBox_ZlJk"></div></div></div><div role="presentation" class="navbar-sidebar__backdrop"></div></nav><div id="__docusaurus_skipToContent_fallback" class="main-wrapper mainWrapper_z2l0 docsWrapper_BCFX"><button aria-label="Scroll back to top" class="clean-btn theme-back-to-top-button backToTopButton_sjWU" type="button"></button><div class="docPage__5DB"><aside class="theme-doc-sidebar-container docSidebarContainer_b6E3"><div class="sidebarViewport_Xe31"><div class="sidebar_njMd"><nav aria-label="Docs sidebar" class="menu thin-scrollbar menu_SIkG"><ul class="theme-doc-sidebar-menu menu__list"><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/BioinfoTech/intro">简介</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret" aria-expanded="false" href="/BioinfoTech/软件安装及环境配置/系统优化">软件安装及环境配置</a></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret menu__link--active" aria-expanded="true" href="/BioinfoTech/常用分析流程及教程/复现">常用分析流程及教程</a></div><ul style="display:block;overflow:visible;height:auto" class="menu__list"><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link menu__link--active" aria-current="page" tabindex="0" href="/BioinfoTech/常用分析流程及教程/复现">文章复现</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-2 menu__list-item"><a class="menu__link" tabindex="0" href="/BioinfoTech/常用分析流程及教程/拟时序分析">单细胞数据（loom）的拟时序分析</a></li></ul></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret" aria-expanded="false" href="/BioinfoTech/提效/Markdown语法教学">提效</a></div></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/BioinfoTech/test">测试</a></li></ul></nav><button type="button" title="Collapse sidebar" aria-label="Collapse sidebar" class="button button--secondary button--outline collapseSidebarButton_PEFL"><svg width="20" height="20" aria-hidden="true" class="collapseSidebarButtonIcon_kv0_"><g fill="#7a7a7a"><path d="M9.992 10.023c0 .2-.062.399-.172.547l-4.996 7.492a.982.982 0 01-.828.454H1c-.55 0-1-.453-1-1 0-.2.059-.403.168-.551l4.629-6.942L.168 3.078A.939.939 0 010 2.528c0-.548.45-.997 1-.997h2.996c.352 0 .649.18.828.45L9.82 9.472c.11.148.172.347.172.55zm0 0"></path><path d="M19.98 10.023c0 .2-.058.399-.168.547l-4.996 7.492a.987.987 0 01-.828.454h-3c-.547 0-.996-.453-.996-1 0-.2.059-.403.168-.551l4.625-6.942-4.625-6.945a.939.939 0 01-.168-.55 1 1 0 01.996-.997h3c.348 0 .649.18.828.45l4.996 7.492c.11.148.168.347.168.55zm0 0"></path></g></svg></button></div></div></aside><main class="docMainContainer_gTbr"><div class="container padding-top--md padding-bottom--lg"><div class="row"><div class="col docItemCol_VOVn"><div class="docItemContainer_Djhp"><article><nav class="theme-doc-breadcrumbs breadcrumbsContainer_Z_bl" aria-label="Breadcrumbs"><ul class="breadcrumbs" itemscope="" itemtype="https://schema.org/BreadcrumbList"><li class="breadcrumbs__item"><a aria-label="Home page" class="breadcrumbs__link" href="/BioinfoTech/"><svg viewBox="0 0 24 24" class="breadcrumbHomeIcon_YNFT"><path d="M10 19v-5h4v5c0 .55.45 1 1 1h3c.55 0 1-.45 1-1v-7h1.7c.46 0 .68-.57.33-.87L12.67 3.6c-.38-.34-.96-.34-1.34 0l-8.36 7.53c-.34.3-.13.87.33.87H5v7c0 .55.45 1 1 1h3c.55 0 1-.45 1-1z" fill="currentColor"></path></svg></a></li><li class="breadcrumbs__item"><span class="breadcrumbs__link">常用分析流程及教程</span><meta itemprop="position" content="1"></li><li itemscope="" itemprop="itemListElement" itemtype="https://schema.org/ListItem" class="breadcrumbs__item breadcrumbs__item--active"><span class="breadcrumbs__link" itemprop="name">文章复现</span><meta itemprop="position" content="2"></li></ul></nav><div class="tocCollapsible_ETCw theme-doc-toc-mobile tocMobile_ITEo"><button type="button" class="clean-btn tocCollapsibleButton_TO0P">On this page</button></div><div class="theme-doc-markdown markdown"><h1>文章复现</h1><blockquote><p>来自 <a href="https://mp.weixin.qq.com/s/khT12emu_AE1KkSsqih1BA" target="_blank" rel="noopener noreferrer">@生信技能树</a></p></blockquote><p>Integration of Single-Cell RNA Sequencing and Bulk RNA Sequencing Data to Establish and Validate a Prognostic Model for Patients With Lung Adenocarcinoma</p><p>全文数据分析，没有湿实验。在 GEO 数据库下载了一份 scRNA 的数据，两份基因芯片的数据，还下载了 TCGA 的数据。研究肺腺癌（LUAD），整合 scRNA-seq 和传统的 RNA-seq 数据来构建 LUAD 患者的预后模型，并采用两个外部验证队列来验证其风险分层能力。</p><p>文章整体思路如下：</p><ol><li>单细胞降维分群注释，找出 tumor 和 normal 间关键细胞类型。单细胞功能富集，拟时序分析，细胞通讯分析</li><li>TCGA 数据进行差异分析，GO，KEGG 功能富集</li><li><strong>TCGA 数据进行 WGCNA 分析，找到 hub gene 后，取 hub gene 和 TCGA DEGs 的交集，进行后续分析</strong></li><li>单因素 Cox 回归鉴定潜在的预后 DEGs，非负矩阵分解进行 sample clustering（亚型），不同 cluster 的生存分析，免疫浸润分析</li><li>单因素 Cox 回归筛选出的 gene 用 LASSO 模型进一步筛选，筛选完后进行多因素 Cox 回归，计算每个样本的 risk score，根据 risk score 分成两类，进行生存分析，将构建的模型应用于两个 GEO 芯片数据，进行同样分析</li><li>高危低危人群与临床信息相关性分析，分析 risk score 是否可作为预后指标，基因突变分析</li><li>ssGSEA 分析，免疫检查点与肿瘤突变负荷分析</li></ol><p><strong>本次复现只是为了复现图表，数据不一定具有生物学意义（大家可以理解为凑图或者灌水）</strong></p><h2 class="anchor anchorWithStickyNavbar_LWe7" id="step-1-单细胞数据处理">Step 1. 单细胞数据处理<a href="#step-1-单细胞数据处理" class="hash-link" aria-label="Direct link to Step 1. 单细胞数据处理" title="Direct link to Step 1. 单细胞数据处理">​</a></h2><p>先下载好数据，一共 4 个样本就新建 4 个文件夹，名字命名成样本名，文件夹内带&#x27;GSM&#x27;开头的数据是下载的原始数据，重新复制一份，把文件名改成标准的 read10x 函数读入格式，一定注意名字不要写错。</p><p><img loading="lazy" alt="image-20230509172810401" src="/BioinfoTech/assets/images/image-20230509172810401-8aada8df4a2a25a7f31d029eb1446fac.png" width="737" height="540" class="img_ev3q"></p><p>一定要保证这 3 个文件同时存在，而且在同一个文件夹下面。</p><p>示例代码是：</p><div class="language-R codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_biex"><pre tabindex="0" class="prism-code language-R codeBlock_bY9V thin-scrollbar"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token plain">rm(list=ls())</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">options(stringsAsFactors = F)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">library(Seurat)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">sce1 &lt;- CreateSeuratObject(Read10X(&#x27;../10x-results/WT/&#x27;), &quot;wt&quot;)</span><br></span></code></pre><div class="buttonGroup__atx"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_eSgA" aria-hidden="true"><svg viewBox="0 0 24 24" class="copyButtonIcon_y97N"><path fill="currentColor" d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg viewBox="0 0 24 24" class="copyButtonSuccessIcon_LjdS"><path fill="currentColor" d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>重点就是 Read10X 函数读取 文件夹路径，比如：<strong>../10x-results/WT/</strong> ，保证文件夹下面有 3 个文件。</p><h3 class="anchor anchorWithStickyNavbar_LWe7" id="11-qc">1.1 QC<a href="#11-qc" class="hash-link" aria-label="Direct link to 1.1 QC" title="Direct link to 1.1 QC">​</a></h3><div class="language-R codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_biex"><pre tabindex="0" class="prism-code language-R codeBlock_bY9V thin-scrollbar"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token plain">library(Seurat)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">library(harmony)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">library(dplyr)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">library(stringr)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">rm(list = ls())</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain"># 多个数据读取与合并</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">rawdata_path &lt;- &#x27;./rawdata&#x27;</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">filename &lt;- list.files(rawdata_path)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">rawdata_path &lt;- paste(rawdata_path,filename,sep = &#x27;/&#x27;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">sceList &lt;- lapply(rawdata_path, function(x){</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  obj &lt;- CreateSeuratObject(counts = Read10X(x),</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                            project = str_split(x,&#x27;/&#x27;)[[1]][3])</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">})</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">names(sceList) &lt;- filename</span><br></span></code></pre><div class="buttonGroup__atx"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_eSgA" aria-hidden="true"><svg viewBox="0 0 24 24" class="copyButtonIcon_y97N"><path fill="currentColor" d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg viewBox="0 0 24 24" class="copyButtonSuccessIcon_LjdS"><path fill="currentColor" d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>这里 sceList 是一个列表，列表里共 4 个元素，每个元素都是一个 seurat 对象，每个元素的名字就是上面的样本名。rawdata<!-- -->_<!-- -->path 就是上面下载的。gz 文件 所在文件夹的路径，Read10X 函数只需要传入文件夹路径就可以自动读入数。
project 参数传入的是一个字符串，一般用样本名即可，这里用 str<!-- -->_<!-- -->split 函数分割路径字符串得到样本名</p><div class="language-R codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_biex"><pre tabindex="0" class="prism-code language-R codeBlock_bY9V thin-scrollbar"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token plain">sce &lt;- merge(sceList[[1]],sceList[-1],add.cell.ids = names(sceList),project = &#x27;luad&#x27;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">sce@meta.data$group &lt;- str_split(sce@meta.data$orig.ident,&#x27;_&#x27;,simplify = T)[,2]</span><br></span></code></pre><div class="buttonGroup__atx"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_eSgA" aria-hidden="true"><svg viewBox="0 0 24 24" class="copyButtonIcon_y97N"><path fill="currentColor" d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg viewBox="0 0 24 24" class="copyButtonSuccessIcon_LjdS"><path fill="currentColor" d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>把 4 个 seurat 对象用 merge 函数合并后，看一下每个细胞的基本信息，这里我根据作者给出的信息，把疾病和正常分成两组，又增加了一列，ade1 是疾病，nor 是正常</p><p><img loading="lazy" alt="image-20230509173051298" src="/BioinfoTech/assets/images/image-20230509173051298-5acf886a9e740674fce3022be1553b5a.png" width="731" height="622" class="img_ev3q"></p><div class="language-R codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_biex"><pre tabindex="0" class="prism-code language-R codeBlock_bY9V thin-scrollbar"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token plain"># 查看线粒体 (MT 开头）、核糖体 (RPS/RPL 开头）、血红细胞所占比例</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">grep(&#x27;^MT&#x27;,x=rownames(sce@assays$RNA@data),value = T)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">grep(&#x27;^RP[SL]&#x27;,x=rownames(sce@assays$RNA@data),value = T)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">grep(&#x27;^HB[^(P)]&#x27;,x=rownames(sce@assays$RNA@data),value = T)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">sce &lt;- PercentageFeatureSet(sce,&#x27;^MT&#x27;,col.name = &#x27;percent_MT&#x27;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">sce &lt;- PercentageFeatureSet(sce,&#x27;^RP[SL]&#x27;,col.name = &#x27;percent_RP&#x27;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">sce &lt;- PercentageFeatureSet(sce,&#x27;^HB[^(P)]&#x27;,col.name = &#x27;percent_HB&#x27;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">VlnPlot(sce,features = &quot;nCount_RNA&quot;,pt.size = 0,y.max = 10000)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">VlnPlot(sce,features = &quot;nFeature_RNA&quot;,pt.size = 0,y.max = 2500)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">VlnPlot(sce,features = &quot;percent_MT&quot;,pt.size = 0)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">VlnPlot(sce,features = &quot;percent_RP&quot;,pt.size = 0)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">VlnPlot(sce,features = &quot;percent_HB&quot;,pt.size = 0,y.max = 0.1)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">VlnPlot(sce,features = c(&quot;nCount_RNA&quot;,&quot;nFeature_RNA&quot;,&quot;percent_MT&quot;),pt.size = 0,group.by = &#x27;orig.ident&#x27;)</span><br></span></code></pre><div class="buttonGroup__atx"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_eSgA" aria-hidden="true"><svg viewBox="0 0 24 24" class="copyButtonIcon_y97N"><path fill="currentColor" d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg viewBox="0 0 24 24" class="copyButtonSuccessIcon_LjdS"><path fill="currentColor" d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>对于每个细胞，我又添加了三列信息：线粒体基因占比，核糖体基因占比，红细胞基因占比。这几列用来过滤数据，这里重新给 sce 赋值是因为，PercentageFeatureSet 函数的返回值还是一个 seurat 对象，和原来的 sce 相比， 就只是在 meta data 里多了一列而已，所以直接重新赋值，把原来的 seurat 对象 sce 给更新掉</p><p><img loading="lazy" alt="image-20230509173202048" src="/BioinfoTech/assets/images/image-20230509173202048-35114e7bdbe24aee691ad42414079c81.png" width="739" height="1264" class="img_ev3q"></p><p>过滤前先 <code>dim(sce)</code> 看下数据，32538 个 gene，13060 个细胞</p><div class="language-R codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_biex"><pre tabindex="0" class="prism-code language-R codeBlock_bY9V thin-scrollbar"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token plain"># 过滤细胞</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">sce &lt;- subset(sce,subset = nCount_RNA&gt;1000 &amp; nFeature_RNA&gt;300 &amp; percent_MT&lt;25)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain"># 过滤基因</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">sce &lt;- sce[rowSums(sce@assays$RNA@counts&gt;0)&gt;3,]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain"># 过滤线粒体、核糖体、血红细胞进占比高的细胞</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">sce &lt;- subset(sce,subset = percent_MT&lt;25 &amp; percent_RP&lt;30 &amp; percent_HB&lt;0.1)</span><br></span></code></pre><div class="buttonGroup__atx"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_eSgA" aria-hidden="true"><svg viewBox="0 0 24 24" class="copyButtonIcon_y97N"><path fill="currentColor" d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg viewBox="0 0 24 24" class="copyButtonSuccessIcon_LjdS"><path fill="currentColor" d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>过滤完后 22249 个 gene，7892 个细胞</p><div class="language-R codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_biex"><pre tabindex="0" class="prism-code language-R codeBlock_bY9V thin-scrollbar"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token plain"># 细胞周期评分</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain"># S.Score 较高的为 S 期，G2M.Score 较高的为 G2M 期，都比较低的为 G1 期</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">s_feature &lt;- cc.genes.updated.2019$s.genes</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">g2m_feature &lt;- cc.genes.updated.2019$g2m.genes</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">sce &lt;- CellCycleScoring(sce,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                        s.features = s_feature,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                        g2m.features = g2m_feature,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                        set.ident = T)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">VlnPlot(sce,features = c(&#x27;S.Score&#x27;,&#x27;G2M.Score&#x27;),group.by = &#x27;orig.ident&#x27;,pt.size = 0)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">saveRDS(sce,&#x27;sce_qc.rds&#x27;)</span><br></span></code></pre><div class="buttonGroup__atx"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_eSgA" aria-hidden="true"><svg viewBox="0 0 24 24" class="copyButtonIcon_y97N"><path fill="currentColor" d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg viewBox="0 0 24 24" class="copyButtonSuccessIcon_LjdS"><path fill="currentColor" d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>再看一下细胞周期，把细胞周期两个评分加到 meta data 里，更新 sce</p><p><img loading="lazy" alt="image-20230509173319035" src="/BioinfoTech/assets/images/image-20230509173319035-ff4d4fabcdf9dd36a81e42957122836b.png" width="739" height="553" class="img_ev3q"></p><h3 class="anchor anchorWithStickyNavbar_LWe7" id="12-用-harmony-去除批次效应">1.2 用 harmony 去除批次效应<a href="#12-用-harmony-去除批次效应" class="hash-link" aria-label="Direct link to 1.2 用 harmony 去除批次效应" title="Direct link to 1.2 用 harmony 去除批次效应">​</a></h3><div class="language-R codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_biex"><pre tabindex="0" class="prism-code language-R codeBlock_bY9V thin-scrollbar"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token plain">library(harmony)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">library(dplyr)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">library(Seurat)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">library(clustree)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">rm(list = ls())</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">sce &lt;- readRDS(&#x27;sce_qc.rds&#x27;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">sce &lt;- NormalizeData(sce,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                     normalization.method = &#x27;LogNormalize&#x27;,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                     scale.factor = 10000)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">sce &lt;- FindVariableFeatures(sce,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                            selection.method = &quot;vst&quot;,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                            nfeatures = 2000)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain"># 默认用 variableFeature 做 scale</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">sce &lt;- ScaleData(sce)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">sce &lt;- RunPCA(sce,features = VariableFeatures(sce))</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">DimPlot(sce,reduction = &#x27;pca&#x27;,group.by = &#x27;group&#x27;)</span><br></span></code></pre><div class="buttonGroup__atx"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_eSgA" aria-hidden="true"><svg viewBox="0 0 24 24" class="copyButtonIcon_y97N"><path fill="currentColor" d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg viewBox="0 0 24 24" class="copyButtonSuccessIcon_LjdS"><path fill="currentColor" d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>从 PCA 降维的结果看，两个组的批次效应不大，均匀的混合在一起，但是这里还是用 harmony 试试</p><p><img loading="lazy" alt="image-20230509173416747" src="/BioinfoTech/assets/images/image-20230509173416747-0ce5d14f50405774552fd04f86510153.png" width="739" height="573" class="img_ev3q"></p><div class="language-R codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_biex"><pre tabindex="0" class="prism-code language-R codeBlock_bY9V thin-scrollbar"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token plain">sce &lt;- RunHarmony(sce,group.by.vars = &#x27;orig.ident&#x27;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">ElbowPlot(sce,reduction = &#x27;harmony&#x27;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">sce &lt;- RunUMAP(sce,dims = 1:10,reduction = &#x27;harmony&#x27;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">sce &lt;- RunTSNE(sce,dims = 1:10,reduction = &#x27;harmony&#x27;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">DimPlot(sce,reduction = &#x27;umap&#x27;,label = T,group.by = &#x27;group&#x27;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">DimPlot(sce,reduction = &#x27;tsne&#x27;,label = T,group.by = &#x27;group&#x27;)</span><br></span></code></pre><div class="buttonGroup__atx"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_eSgA" aria-hidden="true"><svg viewBox="0 0 24 24" class="copyButtonIcon_y97N"><path fill="currentColor" d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg viewBox="0 0 24 24" class="copyButtonSuccessIcon_LjdS"><path fill="currentColor" d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>选择 orig.ident 这一列，去除 4 个样本之间的批次效应。根据肘部图确定降维后的维度。</p><p><img loading="lazy" alt="image-20230509173455682" src="/BioinfoTech/assets/images/image-20230509173455682-40aa0dacb06c001070b07d89ba47fdc3.png" width="752" height="1164" class="img_ev3q"></p><div class="language-R codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_biex"><pre tabindex="0" class="prism-code language-R codeBlock_bY9V thin-scrollbar"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token plain">sce &lt;- FindNeighbors(sce,reduction = &#x27;harmony&#x27;,dims = 1:10)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">sce_res &lt;- FindClusters(sce_res,resolution = 0.8)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain"># 设置不同的分辨率查看分组情况</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">sce_res &lt;- sce</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">for (i in c(0.01, 0.05, 0.1, 0.2, 0.3, 0.5,0.8,1)){</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  sce_res &lt;- FindClusters(sce_res,resolution = i)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">}</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">clustree(sce_res,prefix = &#x27;RNA_snn_res.&#x27;)</span><br></span></code></pre><div class="buttonGroup__atx"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_eSgA" aria-hidden="true"><svg viewBox="0 0 24 24" class="copyButtonIcon_y97N"><path fill="currentColor" d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg viewBox="0 0 24 24" class="copyButtonSuccessIcon_LjdS"><path fill="currentColor" d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>FindNeighbors 函数计算细胞间的距离，FindClusters 函数确定分群结果。上面的 UMAP 看着大概分了 7 坨，但实际上并不是 7 个分群，实际分群数量还是要看 FindClusters 的结果。FindClusters 函数分辨率不同，分群数量会不同，具体分多少需要摸索。这里分辨率选了 0.5，对应第六行蓝色，13 个分群，如果选 0.8 的话，感觉分群数量会突然增加好多，不太合适。</p><p><img loading="lazy" alt="image-20230509173520803" src="/BioinfoTech/assets/images/image-20230509173520803-281922742f22e1f2926660f699e2c749.png" width="748" height="587" class="img_ev3q"></p><div class="language-R codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_biex"><pre tabindex="0" class="prism-code language-R codeBlock_bY9V thin-scrollbar"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token plain">sce &lt;- FindClusters(sce,resolution = 0.5)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">saveRDS(sce,file = &#x27;step2_harmony.rds&#x27;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">DimPlot(sce,reduction = &#x27;umap&#x27;,group.by = &#x27;orig.ident&#x27;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">DimPlot(sce,reduction = &#x27;umap&#x27;,group.by = &#x27;seurat_clusters&#x27;)</span><br></span></code></pre><div class="buttonGroup__atx"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_eSgA" aria-hidden="true"><svg viewBox="0 0 24 24" class="copyButtonIcon_y97N"><path fill="currentColor" d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg viewBox="0 0 24 24" class="copyButtonSuccessIcon_LjdS"><path fill="currentColor" d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>确定好分辨率后，分群赋值给 sce，在 meta data 里会多出来两列，这两列是一样的，表示每个细胞分在哪个 cluster 里，最后保存一下更新后的 sce 对象，再来看下分群结果</p><p><img loading="lazy" alt="image-20230509173607692" src="/BioinfoTech/assets/images/image-20230509173607692-2503c8eadb15233253ad14e753336db1.png" width="748" height="1291" class="img_ev3q"></p><h3 class="anchor anchorWithStickyNavbar_LWe7" id="13-细胞注释">1.3 细胞注释<a href="#13-细胞注释" class="hash-link" aria-label="Direct link to 1.3 细胞注释" title="Direct link to 1.3 细胞注释">​</a></h3><div class="language-R codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_biex"><pre tabindex="0" class="prism-code language-R codeBlock_bY9V thin-scrollbar"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token plain">library(SingleR)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">library(celldex)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">library(Seurat)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">library(dplyr)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">library(stringr)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">library(pheatmap)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">library(ReactomeGSA)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">library(monocle)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">library(limma)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">library(ggplot2)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">library(msigdbr)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">library(singleseqgset)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">rm(list = ls())</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">gc()</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">sce &lt;- readRDS(&#x27;step2_harmony.rds&#x27;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain"># singleR 注释</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">hpca.se &lt;- HumanPrimaryCellAtlasData()</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">bpe.se &lt;- BlueprintEncodeData()</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">anno &lt;- SingleR(sce@assays$RNA@data,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                ref = list(BP=bpe.se,HPCA=hpca.se),</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                labels = list(bpe.se$label.main,hpca.se$label.main),</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                clusters = sce@meta.data$seurat_clusters</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                )</span><br></span></code></pre><div class="buttonGroup__atx"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_eSgA" aria-hidden="true"><svg viewBox="0 0 24 24" class="copyButtonIcon_y97N"><path fill="currentColor" d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg viewBox="0 0 24 24" class="copyButtonSuccessIcon_LjdS"><path fill="currentColor" d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>用 SingleR 自动注释，HumanPrimaryCellAtlasData 和 BlueprintEncodeData 是 SingleR 自带的两个数据库，用两个数据库注释效果好像比一个会好点。得到的 anno 也是一个对象，每个细胞群的注释信息就藏在里面，13 个分群被注释成 7 种细胞，其中有 6 个分群都是上皮细胞，也许上皮细胞还能继续细分。想继续细分的话，对 sce 取子集，把标签为上皮细胞的细胞取出来，降维分群，丢到 SingleR() 函数里再注释</p><div class="language-R codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_biex"><pre tabindex="0" class="prism-code language-R codeBlock_bY9V thin-scrollbar"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token plain">table(anno$labels)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">plotScoreHeatmap(anno,clusters = anno@rownames,show_colnames = T)</span><br></span></code></pre><div class="buttonGroup__atx"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_eSgA" aria-hidden="true"><svg viewBox="0 0 24 24" class="copyButtonIcon_y97N"><path fill="currentColor" d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg viewBox="0 0 24 24" class="copyButtonSuccessIcon_LjdS"><path fill="currentColor" d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>再看下注释结果的打分，不满意的话也可以手动注释</p><p><img loading="lazy" alt="image-20230509173852273" src="/BioinfoTech/assets/images/image-20230509173852273-927e3c0d6c5d7c88557c3228fcfa7d8a.png" width="748" height="467" class="img_ev3q"></p><div class="language-R codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_biex"><pre tabindex="0" class="prism-code language-R codeBlock_bY9V thin-scrollbar"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token plain">sce@meta.data$singleR_label &lt;- unlist(lapply(sce@meta.data$seurat_clusters, function(x){anno$labels[x]}))</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">DimPlot(sce,reduction = &#x27;umap&#x27;,group.by = &#x27;singleR_label&#x27;,label = T)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">DimPlot(sce,reduction = &#x27;tsne&#x27;,group.by = &#x27;singleR_label&#x27;,label = T)</span><br></span></code></pre><div class="buttonGroup__atx"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_eSgA" aria-hidden="true"><svg viewBox="0 0 24 24" class="copyButtonIcon_y97N"><path fill="currentColor" d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg viewBox="0 0 24 24" class="copyButtonSuccessIcon_LjdS"><path fill="currentColor" d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>anno 里面有每个分群对应的注释信息，把这个信息添加到 sce 的 meta data 里，然后就可以在分群的图上添加上注释信息了</p><div class="language-R codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_biex"><pre tabindex="0" class="prism-code language-R codeBlock_bY9V thin-scrollbar"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token plain">anno$labels</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">anno@rownames</span><br></span></code></pre><div class="buttonGroup__atx"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_eSgA" aria-hidden="true"><svg viewBox="0 0 24 24" class="copyButtonIcon_y97N"><path fill="currentColor" d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg viewBox="0 0 24 24" class="copyButtonSuccessIcon_LjdS"><path fill="currentColor" d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p><img loading="lazy" alt="image-20230509174029821" src="/BioinfoTech/assets/images/image-20230509174029821-62217b3f3c4b27910632649dab41794c.png" width="748" height="1164" class="img_ev3q"></p><h3 class="anchor anchorWithStickyNavbar_LWe7" id="14-鉴定关键细胞类型">1.4 鉴定关键细胞类型<a href="#14-鉴定关键细胞类型" class="hash-link" aria-label="Direct link to 1.4 鉴定关键细胞类型" title="Direct link to 1.4 鉴定关键细胞类型">​</a></h3><div class="language-R codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_biex"><pre tabindex="0" class="prism-code language-R codeBlock_bY9V thin-scrollbar"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token plain">cell_list &lt;- split(colnames(sce), sce$singleR_label)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">deg &lt;- c()</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">for (i in names(cell_list)){</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  sce_temp &lt;- sce[,colnames(sce) %in% cell_list[[i]]]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  sce_temp &lt;- CreateSeuratObject(counts = sce_temp@assays$RNA@counts,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                                 meta.data = sce_temp@meta.data,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                                 min.cells = 3,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                                 min.features = 200)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  sce_temp &lt;- NormalizeData(sce_temp)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  sce_temp &lt;- FindVariableFeatures(sce_temp)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  sce_temp &lt;- ScaleData(sce_temp,vars.to.regress = c(&#x27;nFeature_RNA&#x27;,&#x27;&quot;percent_MT&quot;&#x27;))</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  Idents(sce_temp) &lt;- sce_temp$group</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  sce_temp_markers &lt;- FindAllMarkers(object=sce_temp,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                                     only.pos = T,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                                     logfc.threshold = 2,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                                     min.pct = 0.1)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  sce_temp_markers &lt;- sce_temp_markers[order(sce_temp_markers$cluster,sce_temp_markers$avg_log2FC),]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  deg_gene &lt;- sce_temp_markers[sce_temp_markers$p_val_adj&lt;0.05,&#x27;gene&#x27;]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  deg &lt;- c(deg,length(deg_gene))</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">}</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">df &lt;- data.frame(cell_type = names(cell_list),deg_gene = deg)</span><br></span></code></pre><div class="buttonGroup__atx"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_eSgA" aria-hidden="true"><svg viewBox="0 0 24 24" class="copyButtonIcon_y97N"><path fill="currentColor" d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg viewBox="0 0 24 24" class="copyButtonSuccessIcon_LjdS"><path fill="currentColor" d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>把每种类型的细胞分别拿出来，用 FindAllMarkers 找出 ade1 和 nor 组之间的差异基因！<!-- -->[图片]<!-- -->(data:image/svg+xml,%3C%3Fxml version=&#x27;1.0&#x27; encoding=&#x27;UTF-8&#x27;%3F%3E%3Csvg width=&#x27;1px&#x27; height=&#x27;1px&#x27; viewBox=&#x27;0 0 1 1&#x27; version=&#x27;1.1&#x27; xmlns=&#x27;<a href="http://www.w3.org/2000/svg&#x27;" target="_blank" rel="noopener noreferrer">http://www.w3.org/2000/svg&#x27;</a> xmlns:xlink=&#x27;<a href="http://www.w3.org/1999/xlink&#x27;%3E%3Ctitle%3E%3C/title%3E%3Cg" target="_blank" rel="noopener noreferrer">http://www.w3.org/1999/xlink&#x27;%3E%3Ctitle%3E%3C/title%3E%3Cg</a> stroke=&#x27;none&#x27; stroke-width=&#x27;1&#x27; fill=&#x27;none&#x27; fill-rule=&#x27;evenodd&#x27; fill-opacity=&#x27;0&#x27;%3E%3Cg transform=&#x27;translate(-249.000000, -126.000000)&#x27; fill=&#x27;%23FFFFFF&#x27;%3E%3Crect x=&#x27;249&#x27; y=&#x27;126&#x27; width=&#x27;1&#x27; height=&#x27;1&#x27;%3E%3C/rect%3E%3C/g%3E%3C/g%3E%3C/svg%3E)</p><h3 class="anchor anchorWithStickyNavbar_LWe7" id="15-单细胞通路富集">1.5 单细胞通路富集<a href="#15-单细胞通路富集" class="hash-link" aria-label="Direct link to 1.5 单细胞通路富集" title="Direct link to 1.5 单细胞通路富集">​</a></h3><div class="language-R codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_biex"><pre tabindex="0" class="prism-code language-R codeBlock_bY9V thin-scrollbar"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token plain">temp &lt;- sce</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">Idents(temp) &lt;- temp@meta.data$singleR_label</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">reactome_obj &lt;- analyse_sc_clusters(temp,create_reactome_visualization=T,use_interactors=F)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">path_ways &lt;- pathways(reactome_obj)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">path_ways$diff &lt;- rowMax(as.matrix(path_ways[,2:ncol(path_ways)])) - rowMin(as.matrix(path_ways[,2:ncol(path_ways)]))</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">path_ways &lt;- path_ways[order(path_ways$diff,decreasing = T),]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">plot_gsva_heatmap(reactome_obj,rownames(path_ways)[1:10],margins = c(10,20))</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">saveRDS(sce,file = &#x27;step3_sce.rds&#x27;)</span><br></span></code></pre><div class="buttonGroup__atx"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_eSgA" aria-hidden="true"><svg viewBox="0 0 24 24" class="copyButtonIcon_y97N"><path fill="currentColor" d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg viewBox="0 0 24 24" class="copyButtonSuccessIcon_LjdS"><path fill="currentColor" d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>用 pathway() 提取富集出来的结果，每行是一个通路，每列是一种细胞，计算每个通路在 7 种细胞中表达差异的 diff，根据 diff 排序，画出前 10 个通路</p><p><img loading="lazy" alt="image-20230509174111595" src="/BioinfoTech/assets/images/image-20230509174111595-118c2aec5ac7cc1f83012707e80c32b4.png" width="756" height="819" class="img_ev3q"></p><h3 class="anchor anchorWithStickyNavbar_LWe7" id="16-拟时序分析">1.6 拟时序分析<a href="#16-拟时序分析" class="hash-link" aria-label="Direct link to 1.6 拟时序分析" title="Direct link to 1.6 拟时序分析">​</a></h3><div class="language-R codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_biex"><pre tabindex="0" class="prism-code language-R codeBlock_bY9V thin-scrollbar"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token plain">library(Seurat)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">library(monocle)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">sce &lt;- readRDS(&#x27;step3_sce.rds&#x27;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain"># 1. 将 seurat 对象转化为 monocle 的 CDS 对象</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">sparse_data &lt;- as(as.matrix(sce@assays$RNA@counts),&#x27;sparseMatrix&#x27;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">mdata &lt;- new(&#x27;AnnotatedDataFrame&#x27;,data=sce@meta.data) # 行为 cell</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">fData &lt;- data.frame(gene_short_name=row.names(sparse_data),row.names = row.names(sparse_data))</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">fd &lt;- new(&#x27;AnnotatedDataFrame&#x27;,data=fData) # 行为 gene 列为 gene description</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">monocle_cds &lt;- newCellDataSet(cellData = sparse_data,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                              phenoData = mdata,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                              featureData = fd,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                              lowerDetectionLimit = 0.5,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                              expressionFamily = negbinomial.size())</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain"># 2. 计算 size factor 和离散度</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">monocle_cds &lt;- estimateSizeFactors(monocle_cds)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">monocle_cds &lt;- estimateDispersions(monocle_cds)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain"># 3. 过滤低质量细胞</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">monocle_cds &lt;- detectGenes(monocle_cds,min_expr = 0.1)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain"># 4. 找高变基因、降维分群</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain"># 用 seurat 找的高变基因 </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">sce_var_gene &lt;- VariableFeatures(sce)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">monocle_cds &lt;- setOrderingFilter(monocle_cds,sce_var_gene)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">monocle_cds &lt;- reduceDimension(monocle_cds,num_dim=10,norm_method = &#x27;log&#x27;,reduction_method = &#x27;tSNE&#x27;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">monocle_cds &lt;- clusterCells(monocle_cds,num_clusters = 10)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">plot_cell_clusters(monocle_cds,color_by = &#x27;singleR_label&#x27;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">diff_test_gene &lt;- differentialGeneTest(monocle_cds[sce_var_gene,],fullModelFormulaStr = &#x27;~singleR_label&#x27;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">diff_gene &lt;- row.names(subset(diff_test_gene,qval&lt;0.01))</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">monocle_cds &lt;- setOrderingFilter(monocle_cds,diff_gene)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain"># 5. 用 reversed graph embedding 降维</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">monocle_cds &lt;- reduceDimension(monocle_cds,reduction_method = &#x27;DDRTree&#x27;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain"># 6. 计算细胞拟时间</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">monocle_cds &lt;- orderCells(monocle_cds)</span><br></span></code></pre><div class="buttonGroup__atx"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_eSgA" aria-hidden="true"><svg viewBox="0 0 24 24" class="copyButtonIcon_y97N"><path fill="currentColor" d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg viewBox="0 0 24 24" class="copyButtonSuccessIcon_LjdS"><path fill="currentColor" d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>这里运行到第六步，orderCells() 函数大概率会报错，详细的解决办法可以参考知乎上的贴子 <a href="https://zhuanlan.zhihu.com/p/589134519" target="_blank" rel="noopener noreferrer">https://zhuanlan.zhihu.com/p/589134519</a>   ，这里说一下大概解决步骤</p><ol><li><strong>下载一个包，‘igraph’</strong></li><li><strong>下载一个 R 文件，order<!-- -->_<!-- -->cells.R，链接在知乎贴子里，然后 source 调用</strong></li><li><strong>把 orderCells() 中用 DDRTree 计算的那部分代码单独拿出来，写成一个函数，后面直接调用，记得给出返回值。</strong>或者不封装成函数的形式也行，那就不需要返回值了</li></ol><div class="language-R codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_biex"><pre tabindex="0" class="prism-code language-R codeBlock_bY9V thin-scrollbar"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token plain"># 6. 计算细胞拟时间</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">source(&#x27;order_cells.R&#x27;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">library(&#x27;igraph&#x27;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">my_ordercell &lt;- function(cds){</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  root_state = NULL</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  num_paths = NULL</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  reverse = NULL</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  root_cell &lt;- select_root_cell(cds, root_state, reverse)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  cds@auxOrderingData &lt;- new.env(hash = TRUE)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  if (cds@dim_reduce_type == &quot;DDRTree&quot;) {</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    if (is.null(num_paths) == FALSE) {</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">      message(&quot;Warning: num_paths only valid for method &#x27;ICA&#x27; in reduceDimension()&quot;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    }</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    cc_ordering &lt;- extract_ddrtree_ordering(cds, root_cell)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    pData(cds)$Pseudotime &lt;- cc_ordering[row.names(pData(cds)), ]$pseudo_time</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    K_old &lt;- reducedDimK(cds)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    old_dp &lt;- cellPairwiseDistances(cds)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    old_mst &lt;- minSpanningTree(cds)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    old_A &lt;- reducedDimA(cds)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    old_W &lt;- reducedDimW(cds)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    cds &lt;- project2MST(cds, project_point_to_line_segment)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    minSpanningTree(cds) &lt;- cds@auxOrderingData[[cds@dim_reduce_type]]$pr_graph_cell_proj_tree</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    root_cell_idx &lt;- which(V(old_mst)$name == root_cell, arr.ind = T)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    cells_mapped_to_graph_root &lt;- which(cds@auxOrderingData[[&quot;DDRTree&quot;]]$pr_graph_cell_proj_closest_vertex == root_cell_idx)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    if (length(cells_mapped_to_graph_root) == 0) {</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">      cells_mapped_to_graph_root &lt;- root_cell_idx</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    }</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    cells_mapped_to_graph_root &lt;- V(minSpanningTree(cds))[cells_mapped_to_graph_root]$name</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    tip_leaves &lt;- names(which(degree(minSpanningTree(cds)) == 1))</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    root_cell &lt;- cells_mapped_to_graph_root[cells_mapped_to_graph_root %in% tip_leaves][1]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    if (is.na(root_cell)) {</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">      root_cell &lt;- select_root_cell(cds, root_state, reverse)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    }</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    cds@auxOrderingData[[cds@dim_reduce_type]]$root_cell &lt;- root_cell</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    cc_ordering_new_pseudotime &lt;- extract_ddrtree_ordering(cds, root_cell)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    pData(cds)$Pseudotime &lt;- cc_ordering_new_pseudotime[row.names(pData(cds)), ]$pseudo_time</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    if (is.null(root_state) == TRUE) {</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">      closest_vertex &lt;- cds@auxOrderingData[[&quot;DDRTree&quot;]]$pr_graph_cell_proj_closest_vertex</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">      pData(cds)$State &lt;- cc_ordering[closest_vertex[, 1], ]$cell_state</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    }</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  }</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  cds</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">}</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">monocle_cds &lt;- my_ordercell(monocle_cds)</span><br></span></code></pre><div class="buttonGroup__atx"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_eSgA" aria-hidden="true"><svg viewBox="0 0 24 24" class="copyButtonIcon_y97N"><path fill="currentColor" d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg viewBox="0 0 24 24" class="copyButtonSuccessIcon_LjdS"><path fill="currentColor" d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><div class="language-R codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_biex"><pre tabindex="0" class="prism-code language-R codeBlock_bY9V thin-scrollbar"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token plain"># 可以画图了</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">plot_cell_trajectory(monocle_cds,color_by = &#x27;Pseudotime&#x27;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">plot_cell_trajectory(monocle_cds,color_by = &#x27;State&#x27;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">plot_cell_trajectory(monocle_cds,color_by = &#x27;singleR_label&#x27;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">plot_cell_trajectory(monocle_cds,color_by = &#x27;singleR_label&#x27;)+facet_wrap(~singleR_label,nrow = 3)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">saveRDS(monocle_cds,file = &#x27;monocle.rds&#x27;)</span><br></span></code></pre><div class="buttonGroup__atx"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_eSgA" aria-hidden="true"><svg viewBox="0 0 24 24" class="copyButtonIcon_y97N"><path fill="currentColor" d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg viewBox="0 0 24 24" class="copyButtonSuccessIcon_LjdS"><path fill="currentColor" d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p><img loading="lazy" alt="image-20230509174211956" src="/BioinfoTech/assets/images/image-20230509174211956-7449305127e498f54bba72e7b0b8d1f4.png" width="756" height="1140" class="img_ev3q"></p><h3 class="anchor anchorWithStickyNavbar_LWe7" id="17-细胞通讯">1.7 细胞通讯<a href="#17-细胞通讯" class="hash-link" aria-label="Direct link to 1.7 细胞通讯" title="Direct link to 1.7 细胞通讯">​</a></h3><div class="language-R codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_biex"><pre tabindex="0" class="prism-code language-R codeBlock_bY9V thin-scrollbar"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token plain">library(CellChat)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">library(patchwork)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">library(Seurat)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">library(dplyr)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">library(ggalluvial)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">library(NMF)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">options(stringsAsFactors = FALSE)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">rm(list = ls())</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">gc()</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">sce &lt;- readRDS(&#x27;step3_sce.rds&#x27;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain"># Step1. 创建 cellchat 对象</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">data.input &lt;- sce@assays$RNA@data</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">meta.data &lt;- sce@meta.data</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">cellchat &lt;- createCellChat(object=data.input,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                           meta = meta.data,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                           group.by=&#x27;singleR_label&#x27;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">cellchat &lt;- addMeta(cellchat,meta = meta.data)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain"># Step2. 加载 CellChatDB 数据库</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">cellchatDB &lt;- CellChatDB.human</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">cellchat@DB &lt;- cellchatDB</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain"># Step3. 处理表达数据</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">cellchat &lt;- subsetData(cellchat)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">cellchat &lt;- identifyOverExpressedGenes(cellchat)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">cellchat &lt;- identifyOverExpressedInteractions(cellchat)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain"># Step4. 计算通讯概率，推断细胞通讯网络  这一步很耗时</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">cellchat &lt;- computeCommunProb(cellchat,population.size = F)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">cellchat &lt;- filterCommunication(cellchat,min.cells = 10)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain"># Step5. 提取预测的细胞通讯网络为 data frame</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">df.net &lt;- subsetCommunication(cellchat)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">df.pathway &lt;- subsetCommunication(cellchat,slot.name = &#x27;netP&#x27;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain"># Step6. 在信号通路水平推断细胞通讯</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">cellchat &lt;- computeCommunProbPathway(cellchat)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain"># Step7. 计算加和的 cell-cell 通讯网络</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">cellchat &lt;- aggregateNet(cellchat)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">par(mfrow = c(1,2))</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">netVisual_circle(cellchat@net$count,vertex.weight = groupSize,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                 weight.scale = T,label.edge = F,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                 title.name = &#x27;number of Interaction&#x27;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">netVisual_circle(cellchat@net$weight,vertex.weight = groupSize,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                 weight.scale = T,label.edge = F,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                 title.name = &#x27;Interaction Weight&#x27;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">mat &lt;- cellchat@net$weight</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">par(mfrow = c(3,3),xpd=T)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">for (i in 1:nrow(mat)){</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  mat2 &lt;- matrix(0,nrow = nrow(mat),ncol = ncol(mat),</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                 dimnames = dimnames(mat))</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  mat2[i,] &lt;- mat[i,]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  netVisual_circle(mat2,vertex.weight = groupSize,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                   weight.scale = T,edge.weight.max = max(mat),</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                   title.name = rownames(mat)[i])</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">}</span><br></span></code></pre><div class="buttonGroup__atx"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_eSgA" aria-hidden="true"><svg viewBox="0 0 24 24" class="copyButtonIcon_y97N"><path fill="currentColor" d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg viewBox="0 0 24 24" class="copyButtonSuccessIcon_LjdS"><path fill="currentColor" d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p><img loading="lazy" alt="image-20230509175412678" src="/BioinfoTech/assets/images/image-20230509175412678-9035d2cb8394bfc08a42636875b0dc7f.png" width="756" height="304" class="img_ev3q"></p><div class="language-R codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_biex"><pre tabindex="0" class="prism-code language-R codeBlock_bY9V thin-scrollbar"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token plain"># Step8. 使用层次图（Hierarchical plot），圆圈图（Circle plot）或和弦图（Chord diagram）可视化每个信号通路</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">cellchat@netP$pathways</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">pathway.show &lt;- c(&#x27;SEMA4&#x27;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">vertex.receiver &lt;- c(1,2,3,4)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">netVisual_aggregate(cellchat,signaling = pathway.show,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                    vertex.receiver = vertex.receiver,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                    layout = &#x27;hierarchy&#x27;)</span><br></span></code></pre><div class="buttonGroup__atx"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_eSgA" aria-hidden="true"><svg viewBox="0 0 24 24" class="copyButtonIcon_y97N"><path fill="currentColor" d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg viewBox="0 0 24 24" class="copyButtonSuccessIcon_LjdS"><path fill="currentColor" d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p><img loading="lazy" alt="image-20230509175509671" src="/BioinfoTech/assets/images/image-20230509175509671-4b5b57a54fd4ce082b707234ec811798.png" width="756" height="309" class="img_ev3q"></p><div class="language-R codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_biex"><pre tabindex="0" class="prism-code language-R codeBlock_bY9V thin-scrollbar"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token plain">par(mfrow=c(1,2),xpd=T)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">netVisual_aggregate(cellchat,signaling = pathway.show,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                    layout = &#x27;circle&#x27;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">netVisual_aggregate(cellchat,signaling = pathway.show,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                    layout = &#x27;circle&#x27;,label.edge=T)</span><br></span></code></pre><div class="buttonGroup__atx"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_eSgA" aria-hidden="true"><svg viewBox="0 0 24 24" class="copyButtonIcon_y97N"><path fill="currentColor" d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg viewBox="0 0 24 24" class="copyButtonSuccessIcon_LjdS"><path fill="currentColor" d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p><img loading="lazy" alt="image-20230509175527871" src="/BioinfoTech/assets/images/image-20230509175527871-884cb17837749eeffc3a0f71f90ea94b.png" width="756" height="305" class="img_ev3q"></p><div class="language-R codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_biex"><pre tabindex="0" class="prism-code language-R codeBlock_bY9V thin-scrollbar"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token plain">pairLR.CXCL &lt;- extractEnrichedLR(cellchat,signaling = pathway.show,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                                 geneLR.return = F)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">LR.show &lt;- pairLR.CXCL[1,]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">netVisual_individual(cellchat,signaling = pathway.show,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                     pairLR.use = &#x27;SEMA4A_PLXNB2&#x27;,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                     layout = &#x27;circle&#x27;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">netVisual_aggregate(cellchat,signaling = pathway.show,layout = &#x27;chord&#x27;,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                    title.name = &quot;Chord diagram  1&quot;)</span><br></span></code></pre><div class="buttonGroup__atx"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_eSgA" aria-hidden="true"><svg viewBox="0 0 24 24" class="copyButtonIcon_y97N"><path fill="currentColor" d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg viewBox="0 0 24 24" class="copyButtonSuccessIcon_LjdS"><path fill="currentColor" d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p><img loading="lazy" alt="image-20230509175623495" src="/BioinfoTech/assets/images/image-20230509175623495-93f9f0ae41db08ca6f7e5fe0472795f3.png" width="690" height="1333" class="img_ev3q"></p><h2 class="anchor anchorWithStickyNavbar_LWe7" id="step-2-tcga-差异分析">Step 2. TCGA 差异分析<a href="#step-2-tcga-差异分析" class="hash-link" aria-label="Direct link to Step 2. TCGA 差异分析" title="Direct link to Step 2. TCGA 差异分析">​</a></h2><div class="language-R codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_biex"><pre tabindex="0" class="prism-code language-R codeBlock_bY9V thin-scrollbar"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token plain">library(TCGAbiolinks)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">library(SummarizedExperiment)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">library(AnnoProbe)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">library(stringr)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">library(limma)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">library(tinyarray)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">library(data.table)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">library(ggplot2)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">library(dplyr)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">library(tibble)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">library(pheatmap)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">library(clusterProfiler)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">library(org.Hs.eg.db)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">rm(list = ls())</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain"># 下载 LUAD 表达矩阵</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">query &lt;- GDCquery(project = &#x27;TCGA-LUAD&#x27;,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                  data.category = &#x27;Transcriptome Profiling&#x27;,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                  data.type = &#x27;Gene Expression Quantification&#x27;,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                  workflow.type = &#x27;STAR - Counts&#x27;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">GDCdownload(query)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">dat &lt;- GDCprepare(query)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">exp &lt;- assay(dat)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">save(exp,file=&#x27;TCGA_LUAD_exp.Rdata&#x27;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain"># 下载 LUAD 临床信息</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">query &lt;- GDCquery(project = &#x27;TCGA-LUAD&#x27;,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                  data.category = &#x27;Clinical&#x27;,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                  data.type = &#x27;Clinical Supplement&#x27;,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                  file.type = &#x27;xml&#x27;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">GDCdownload(query)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">dat &lt;- GDCprepare_clinic(query,clinical.info = &#x27;patient&#x27;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">k = apply(dat, 2, function(x){!all(is.na(x))});table(k)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">clinical &lt;- dat[,k]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">save(clinical,file = &#x27;TCGA_LUAD_Clinical.Rdata&#x27;)</span><br></span></code></pre><div class="buttonGroup__atx"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_eSgA" aria-hidden="true"><svg viewBox="0 0 24 24" class="copyButtonIcon_y97N"><path fill="currentColor" d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg viewBox="0 0 24 24" class="copyButtonSuccessIcon_LjdS"><path fill="currentColor" d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>注意一下 TCGA 的样本命名含义： </p><p>以这个为例：TCGA-A6-6650-01A-11R-1774-07</p><p><strong>A6</strong>：Tissue source site，组织来源编码</p><p><strong>6650</strong>：Participant, 参与者编号</p><p><strong>01</strong>：Sample, 编号 01~09 表示肿瘤，10~19 表示正常对照</p><p><strong>A</strong>：Vial, 在一系列患者组织中的顺序，绝大多数样本该位置编码都是 A; 很少数的是 B，表示福尔马林固定石蜡包埋组织，已被证明用于测序分析的效果不佳，所以不建议使用-01B 的样本数据</p><p><strong>11</strong>：Portion, 同属于一个患者组织的不同部分的顺序编号，同一组织会分割为 100-120mg 的部分，分别使用</p><p><strong>R</strong>：RNA, 分析的分子类型，对应关系如下所示：</p><p><strong>1774</strong>：Plate, 在一系列 96 孔板中的顺序，值大表示制板越晚</p><p><strong>07</strong>：Center, 测序或鉴定中心编码</p><div class="language-R codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_biex"><pre tabindex="0" class="prism-code language-R codeBlock_bY9V thin-scrollbar"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token plain"># count 转换成 TPM</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">gene_length &lt;- fread(&#x27;gene_len.txt&#x27;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">colnames(gene_length) &lt;- c(&#x27;id&#x27;,&#x27;length&#x27;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">gene_length &lt;- as.data.frame(gene_length)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">gene_length &lt;- gene_length[!duplicated(gene_length$id),]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">exp &lt;- as.data.frame(exp)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">exp$id &lt;- unlist(lapply(rownames(exp),function(x){str_split(x,&#x27;\\.&#x27;,simplify = T)[,1]}))</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">merge &lt;- left_join(x=exp,y=gene_length,by=&#x27;id&#x27;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">merge &lt;- na.omit(merge)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">merge &lt;- merge[!duplicated(merge$id),]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">rownames(merge) &lt;- merge$id</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">merge &lt;- merge[,-601]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain"># 计算 TPM</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">kb &lt;- merge$length / 1000</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">rpk &lt;- merge[,-601] / kb</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">tpm &lt;- t(t(rpk) / colSums(rpk) * 1E6)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">exp &lt;- tpm</span><br></span></code></pre><div class="buttonGroup__atx"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_eSgA" aria-hidden="true"><svg viewBox="0 0 24 24" class="copyButtonIcon_y97N"><path fill="currentColor" d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg viewBox="0 0 24 24" class="copyButtonSuccessIcon_LjdS"><path fill="currentColor" d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>采用 limma 做差异分析，先把 counts 转换成 TPM。不转也可以用 DEseq2 做差异分析。gene<!-- -->_<!-- -->length 是每个基因的长度，这个文件是把 gtf 文件中，基因的终止位置-起始位置得到的。</p><div class="language-R codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_biex"><pre tabindex="0" class="prism-code language-R codeBlock_bY9V thin-scrollbar"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token plain"># 更改行名：ENSEMBL -&gt; SYMBOL</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">a &lt;- annoGene(str_split(rownames(exp),&#x27;\\.&#x27;,simplify = T)[,1],&#x27;ENSEMBL&#x27;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">a &lt;- a[!duplicated(a$ENSEMBL),]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">rownames(a) &lt;- a$ENSEMBL</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">rownames(exp) &lt;- a[str_split(rownames(exp),&#x27;\\.&#x27;,simplify = T)[,1],&#x27;SYMBOL&#x27;]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain"># 给样本分组 tumor/normal</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">group &lt;- ifelse(sapply(str_split(colnames(exp),&#x27;-&#x27;,simplify = T)[,4],function(x){as.integer(substr(x,1,2))})&gt;=10,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                &#x27;normal&#x27;,&#x27;tumor&#x27;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">sample_group &lt;- data.frame(sample=colnames(exp),group=group)</span><br></span></code></pre><div class="buttonGroup__atx"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_eSgA" aria-hidden="true"><svg viewBox="0 0 24 24" class="copyButtonIcon_y97N"><path fill="currentColor" d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg viewBox="0 0 24 24" class="copyButtonSuccessIcon_LjdS"><path fill="currentColor" d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>a 是 annoGene 函数返回的一个 ID 转换的结果，是一个 dataframe，多个 ENSEMBL ID 可能对应同一个 SYMBOL ID，对 symbol 去重后，再映射成 exp 的行名。</p><p><img loading="lazy" alt="image-20230509175922794" src="/BioinfoTech/assets/images/image-20230509175922794-e15b8baba93868af1bdfc325292cf67e.png" width="752" height="311" class="img_ev3q"></p><p>按照 TCGA 名字中第四部分，把样本分成 tumor 和 normal</p><p><img loading="lazy" alt="image-20230509175955369" src="/BioinfoTech/assets/images/image-20230509175955369-fd70e0ed205ba80932727543ed5328d5.png" width="752" height="286" class="img_ev3q"></p><div class="language-R codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_biex"><pre tabindex="0" class="prism-code language-R codeBlock_bY9V thin-scrollbar"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token plain"># 过滤</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain"># 1. 去除 FFPE 样本</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">sample_group$sample_type &lt;- ifelse(sapply(str_split(colnames(exp),&#x27;-&#x27;,simplify = T)[,4],function(x){substr(x,3,3)})==&#x27;A&#x27;,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                                   &#x27;tissue&#x27;,&#x27;ffpe&#x27;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">exp &lt;- exp[,sample_group$sample_type!=&#x27;ffpe&#x27;]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">sample_group &lt;- sample_group[sample_group$sample_type==&#x27;tissue&#x27;,]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain"># 2. 保留在一半以上样本中有值的 gene</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">exp &lt;- exp[rowSums(exp&gt;0)&gt;as.integer(ncol(exp)/2),]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">dim(exp)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">save(exp,file = &#x27;TCGA_LUAD_TPM_exp.Rdata&#x27;)</span><br></span></code></pre><div class="buttonGroup__atx"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_eSgA" aria-hidden="true"><svg viewBox="0 0 24 24" class="copyButtonIcon_y97N"><path fill="currentColor" d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg viewBox="0 0 24 24" class="copyButtonSuccessIcon_LjdS"><path fill="currentColor" d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>FFPE 样本无论做测序还是蛋白组，代谢组，效果都不怎么好</p><div class="language-R codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_biex"><pre tabindex="0" class="prism-code language-R codeBlock_bY9V thin-scrollbar"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token plain"># 对表达矩阵取 log2</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">for (i in 1:ncol(exp)){</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  exp[,i] &lt;- log2(exp[,i]+0.0001)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">}</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">dim(exp)</span><br></span></code></pre><div class="buttonGroup__atx"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_eSgA" aria-hidden="true"><svg viewBox="0 0 24 24" class="copyButtonIcon_y97N"><path fill="currentColor" d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg viewBox="0 0 24 24" class="copyButtonSuccessIcon_LjdS"><path fill="currentColor" d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><div class="language-R codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_biex"><pre tabindex="0" class="prism-code language-R codeBlock_bY9V thin-scrollbar"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token plain"># 差异筛选</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">design &lt;- model.matrix(~factor(sample_group$group,levels = c(&#x27;tumor&#x27;,&#x27;normal&#x27;)))</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">fit &lt;- lmFit(exp,design = design)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">fit &lt;- eBayes(fit)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">deg &lt;- topTable(fit, coef = 2,number = Inf)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">deg$change &lt;- ifelse((deg$logFC&gt;log2(2))&amp;(deg$adj.P.Val&lt;0.05),&#x27;up&#x27;,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                     ifelse((deg$logFC&lt;log2(0.5))&amp;(deg$adj.P.Val&lt;0.05),&#x27;down&#x27;,&#x27;nochange&#x27;))</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">deg &lt;- na.omit(deg)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">table(deg$change)</span><br></span></code></pre><div class="buttonGroup__atx"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_eSgA" aria-hidden="true"><svg viewBox="0 0 24 24" class="copyButtonIcon_y97N"><path fill="currentColor" d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg viewBox="0 0 24 24" class="copyButtonSuccessIcon_LjdS"><path fill="currentColor" d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>![图片]<!-- -->(data:image/svg+xml,%3C%3Fxml version=&#x27;1.0&#x27; encoding=&#x27;UTF-8&#x27;%3F%3E%3Csvg width=&#x27;1px&#x27; height=&#x27;1px&#x27; viewBox=&#x27;0 0 1 1&#x27; version=&#x27;1.1&#x27; xmlns=&#x27;<a href="http://www.w3.org/2000/svg&#x27;" target="_blank" rel="noopener noreferrer">http://www.w3.org/2000/svg&#x27;</a> xmlns:xlink=&#x27;<a href="http://www.w3.org/1999/xlink&#x27;%3E%3Ctitle%3E%3C/title%3E%3Cg" target="_blank" rel="noopener noreferrer">http://www.w3.org/1999/xlink&#x27;%3E%3Ctitle%3E%3C/title%3E%3Cg</a> stroke=&#x27;none&#x27; stroke-width=&#x27;1&#x27; fill=&#x27;none&#x27; fill-rule=&#x27;evenodd&#x27; fill-opacity=&#x27;0&#x27;%3E%3Cg transform=&#x27;translate(-249.000000, -126.000000)&#x27; fill=&#x27;%23FFFFFF&#x27;%3E%3Crect x=&#x27;249&#x27; y=&#x27;126&#x27; width=&#x27;1&#x27; height=&#x27;1&#x27;%3E%3C/rect%3E%3C/g%3E%3C/g%3E%3C/svg%3E)<!-- -->![图片]<!-- -->(data:image/svg+xml,%3C%3Fxml version=&#x27;1.0&#x27; encoding=&#x27;UTF-8&#x27;%3F%3E%3Csvg width=&#x27;1px&#x27; height=&#x27;1px&#x27; viewBox=&#x27;0 0 1 1&#x27; version=&#x27;1.1&#x27; xmlns=&#x27;<a href="http://www.w3.org/2000/svg&#x27;" target="_blank" rel="noopener noreferrer">http://www.w3.org/2000/svg&#x27;</a> xmlns:xlink=&#x27;<a href="http://www.w3.org/1999/xlink&#x27;%3E%3Ctitle%3E%3C/title%3E%3Cg" target="_blank" rel="noopener noreferrer">http://www.w3.org/1999/xlink&#x27;%3E%3Ctitle%3E%3C/title%3E%3Cg</a> stroke=&#x27;none&#x27; stroke-width=&#x27;1&#x27; fill=&#x27;none&#x27; fill-rule=&#x27;evenodd&#x27; fill-opacity=&#x27;0&#x27;%3E%3Cg transform=&#x27;translate(-249.000000, -126.000000)&#x27; fill=&#x27;%23FFFFFF&#x27;%3E%3Crect x=&#x27;249&#x27; y=&#x27;126&#x27; width=&#x27;1&#x27; height=&#x27;1&#x27;%3E%3C/rect%3E%3C/g%3E%3C/g%3E%3C/svg%3E) 做完差异筛选，有 10 个 gene 含空值，删去有空值的行即可。更新下 exp 并保存好</p><div class="language-R codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_biex"><pre tabindex="0" class="prism-code language-R codeBlock_bY9V thin-scrollbar"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token plain"># deg 比 exp 少了几个 gene</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">exp &lt;- exp[intersect(rownames(exp),deg$ID),]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">save(exp,file = &#x27;TCGA_exp.Rdata&#x27;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">save(deg,file = &#x27;TCGA_DEGS.Rdata&#x27;)</span><br></span></code></pre><div class="buttonGroup__atx"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_eSgA" aria-hidden="true"><svg viewBox="0 0 24 24" class="copyButtonIcon_y97N"><path fill="currentColor" d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg viewBox="0 0 24 24" class="copyButtonSuccessIcon_LjdS"><path fill="currentColor" d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><div class="language-R codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_biex"><pre tabindex="0" class="prism-code language-R codeBlock_bY9V thin-scrollbar"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token plain"># volcano plot</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">ggplot(deg,aes(x=deg$logFC,y=-log10(deg$adj.P.Val),color=change))+</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  geom_point()+</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  geom_hline(yintercept = -log10(0.05),linetype=2,linewidth=1)+</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  geom_vline(xintercept = log2(0.5),linetype=2,linewidth=1)+</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  geom_vline(xintercept = log2(2),linetype=2,linewidth=1)+</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  scale_color_manual(values=c(&#x27;blue&#x27;,&#x27;gray&#x27;,&#x27;red&#x27;))</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain"># heatmap</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">deg_heatmap &lt;- deg[order(deg$logFC,decreasing = T),]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">deg_gene &lt;- deg_heatmap[deg_heatmap$change!=&#x27;nochange&#x27;,&#x27;ID&#x27;]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">heatmap_deg_gene &lt;- as.data.frame(exp)[deg_gene,]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">heatmap_deg_gene &lt;- na.omit(heatmap_deg_gene)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">anno_col &lt;- data.frame(group=sample_group$group)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">rownames(anno_col) &lt;- sample_group$sample</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">pheatmap(heatmap_deg_gene,show_rownames = F,cluster_rows = F,scale = &#x27;row&#x27;,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">         show_colnames = F,annotation_col=anno_col ,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">         breaks = seq(-2.5,2.5,length.out=100))</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">dim(heatmap_deg_gene)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain"># pca plot</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">exp &lt;- as.data.frame(exp)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">pca &lt;- prcomp(t(exp))</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">ggplot(data.frame(pca$x),aes(x=PC1,y=PC2,color=sample_group$group))+</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  geom_point(size = 3)+stat_ellipse(level = 0.95, show.legend = F)+ </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  labs(title = &#x27;PCA Analysis&#x27;)+</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  theme(plot.title = element_text(hjust = 0.5))+</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  guides(color = guide_legend(title = &#x27;Group&#x27;))</span><br></span></code></pre><div class="buttonGroup__atx"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_eSgA" aria-hidden="true"><svg viewBox="0 0 24 24" class="copyButtonIcon_y97N"><path fill="currentColor" d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg viewBox="0 0 24 24" class="copyButtonSuccessIcon_LjdS"><path fill="currentColor" d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p><img loading="lazy" alt="image-20230509180121206" src="/BioinfoTech/assets/images/image-20230509180121206-b576987f5ca1ed12cae42eff7e82cfcf.png" width="752" height="537" class="img_ev3q"></p><p><img loading="lazy" alt="image-20230509180142980" src="/BioinfoTech/assets/images/image-20230509180142980-c391838fc7652f0e27bf0b0f0cb3d832.png" width="752" height="1070" class="img_ev3q"></p><div class="language-R codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_biex"><pre tabindex="0" class="prism-code language-R codeBlock_bY9V thin-scrollbar"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token plain"># GO enrich</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">go_bp &lt;- enrichGO(deg_gene,org.Hs.eg.db,&#x27;SYMBOL&#x27;,&#x27;BP&#x27;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">go_mf &lt;- enrichGO(deg_gene,org.Hs.eg.db,&#x27;SYMBOL&#x27;,&#x27;MF&#x27;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">go_cc &lt;- enrichGO(deg_gene,org.Hs.eg.db,&#x27;SYMBOL&#x27;,&#x27;CC&#x27;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">dotplot(go_bp)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">barplot(go_bp,showCategory = 4)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain"># KEGG enrich</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">entrez_id &lt;- bitr(deg_gene,fromType = &#x27;SYMBOL&#x27;,toType = &#x27;ENTREZID&#x27;,OrgDb = org.Hs.eg.db)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">kegg &lt;- enrichKEGG(entrez_id$ENTREZID,organism = &#x27;hsa&#x27;,keyType = &#x27;ncbi-geneid&#x27;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">dotplot(kegg)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">barplot(kegg,showCategory = 4)</span><br></span></code></pre><div class="buttonGroup__atx"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_eSgA" aria-hidden="true"><svg viewBox="0 0 24 24" class="copyButtonIcon_y97N"><path fill="currentColor" d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg viewBox="0 0 24 24" class="copyButtonSuccessIcon_LjdS"><path fill="currentColor" d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p><img loading="lazy" alt="image-20230509180204539" src="/BioinfoTech/assets/images/image-20230509180204539-3f674c2271ecdfae5b92eb77c7182238.png" width="752" height="1036" class="img_ev3q"></p><h2 class="anchor anchorWithStickyNavbar_LWe7" id="step-3-wgcna-分析">Step 3. WGCNA 分析<a href="#step-3-wgcna-分析" class="hash-link" aria-label="Direct link to Step 3. WGCNA 分析" title="Direct link to Step 3. WGCNA 分析">​</a></h2><div class="language-R codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_biex"><pre tabindex="0" class="prism-code language-R codeBlock_bY9V thin-scrollbar"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token plain">library(WGCNA)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">library(reshape2)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">library(stringr)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">library(dplyr)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">library(patchwork)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">library(ggplot2)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">rm(list = ls())</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">gc()</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">load(&#x27;TCGA_DEGS.Rdata&#x27;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">load(&#x27;TCGA_exp.Rdata&#x27;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">load(&#x27;TCGA_LUAD_Clinical.Rdata&#x27;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">options(stringsAsFactors = F) # 开启多线程</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">enableWGCNAThreads()</span><br></span></code></pre><div class="buttonGroup__atx"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_eSgA" aria-hidden="true"><svg viewBox="0 0 24 24" class="copyButtonIcon_y97N"><path fill="currentColor" d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg viewBox="0 0 24 24" class="copyButtonSuccessIcon_LjdS"><path fill="currentColor" d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><div class="language-R codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_biex"><pre tabindex="0" class="prism-code language-R codeBlock_bY9V thin-scrollbar"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token plain"># 1. 数据处理</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain"># &gt;&gt;&gt; 选取表达量前 10000 的 gene，用全部 gene 的话，计算后面的邻接矩阵和 TOM 矩阵太慢了</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">temp &lt;- as.data.frame(exp)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">temp$avg_exp &lt;- rowMeans(temp)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">temp &lt;- temp[order(temp$avg_exp,decreasing = T),]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">top10000_gene &lt;- rownames(temp[1:10000,])</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">exp &lt;- as.data.frame(exp[top10000_gene,])</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain"># &gt;&gt;&gt; 查找并删除异常值</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">temp &lt;- exp</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">colnames(temp) &lt;- 1:ncol(exp)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">plot(hclust(dist(t(temp))))  # 没有异常样本</span><br></span></code></pre><div class="buttonGroup__atx"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_eSgA" aria-hidden="true"><svg viewBox="0 0 24 24" class="copyButtonIcon_y97N"><path fill="currentColor" d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg viewBox="0 0 24 24" class="copyButtonSuccessIcon_LjdS"><path fill="currentColor" d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p><img loading="lazy" alt="image-20230509180239398" src="/BioinfoTech/assets/images/image-20230509180239398-8e87c601259b13ab1f8f508d6c9b0a5c.png" width="752" height="426" class="img_ev3q"></p><div class="language-R codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_biex"><pre tabindex="0" class="prism-code language-R codeBlock_bY9V thin-scrollbar"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token plain"># 2. 构建基因共表达网络</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">exp &lt;- as.data.frame(t(exp))</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain"># &gt;&gt;&gt; 2.1 选择合适的 power</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">power_vec &lt;- c(seq(1, 10, by = 1), seq(12, 20, by = 2))</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">sft &lt;- pickSoftThreshold(exp,powerVector = power_vec) </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">a &lt;- ggplot(sft$fitIndices,aes(x=Power,y=-sign(slope)*SFT.R.sq))+</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  geom_text(label=sft$fitIndices$Power,color=&#x27;red&#x27;)+</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  geom_hline(yintercept = 0.9,color=&#x27;red&#x27;)+</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  xlab(&quot;Soft Threshold (power)&quot;)+ylab(&quot;Scale Free Topology Model Fit,signed R^2&quot;)+</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  ggtitle(&quot;Scale independence&quot;)+theme(plot.title = element_text(hjust = 0.5))</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">b &lt;- ggplot(sft$fitIndices,aes(x=Power,y=mean.k.))+</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  geom_text(label=sft$fitIndices$Power,color=&#x27;red&#x27;)+</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  xlab(&quot;Soft Threshold (power)&quot;)+ylab(&quot;Mean Connectivity&quot;)+</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  ggtitle(&#x27;Mean connectivity&#x27;)+theme(plot.title = element_text(hjust = 0.5))</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">a+b</span><br></span></code></pre><div class="buttonGroup__atx"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_eSgA" aria-hidden="true"><svg viewBox="0 0 24 24" class="copyButtonIcon_y97N"><path fill="currentColor" d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg viewBox="0 0 24 24" class="copyButtonSuccessIcon_LjdS"><path fill="currentColor" d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p><img loading="lazy" alt="image-20230509180259854" src="/BioinfoTech/assets/images/image-20230509180259854-e3e5b5830f42c1fa3d1c16824d7e0ba9.png" width="752" height="530" class="img_ev3q"></p><p>这里 power=12，达到 0.8 即可</p><div class="language-R codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_biex"><pre tabindex="0" class="prism-code language-R codeBlock_bY9V thin-scrollbar"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token plain">power &lt;- sft$powerEstimate</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain"># &gt;&gt;&gt; 2.2 检验所选 power 是否符合无尺度网络</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">k &lt;- softConnectivity(exp,power = power)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">scaleFreePlot(k)</span><br></span></code></pre><div class="buttonGroup__atx"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_eSgA" aria-hidden="true"><svg viewBox="0 0 24 24" class="copyButtonIcon_y97N"><path fill="currentColor" d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg viewBox="0 0 24 24" class="copyButtonSuccessIcon_LjdS"><path fill="currentColor" d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>power=12 确实也符合无尺度网络，R2 较高</p><p><img loading="lazy" alt="image-20230509180315053" src="/BioinfoTech/assets/images/image-20230509180315053-186633e8cf3a5b47335643b75673cef7.png" width="752" height="506" class="img_ev3q"></p><div class="language-R codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_biex"><pre tabindex="0" class="prism-code language-R codeBlock_bY9V thin-scrollbar"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token plain"># &gt;&gt;&gt; 2.3 计算 gene 在 sample 间的 corr 或者 distance，构建临近矩阵。默认用 dist 函数计算（欧氏距离）</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">adj &lt;- adjacency(exp,power = power)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain"># &gt;&gt;&gt; 2.4 根据临近矩阵计算 TOM（拓扑重叠矩阵）</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">TOM &lt;- TOMsimilarity(adj)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain"># &gt;&gt;&gt; 2.5 计算 gene 间相异性</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">dissTOM &lt;- 1-TOM</span><br></span></code></pre><div class="buttonGroup__atx"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_eSgA" aria-hidden="true"><svg viewBox="0 0 24 24" class="copyButtonIcon_y97N"><path fill="currentColor" d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg viewBox="0 0 24 24" class="copyButtonSuccessIcon_LjdS"><path fill="currentColor" d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><div class="language-R codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_biex"><pre tabindex="0" class="prism-code language-R codeBlock_bY9V thin-scrollbar"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token plain"># 3. 对 gene 进行聚类并可视化</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">geneTree &lt;- hclust(as.dist(dissTOM),method = &#x27;average&#x27;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">sizeGrWindow(12,9)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">plot(geneTree, xlab=&quot;&quot;, sub=&quot;&quot;, main = &quot;Gene clustering on TOM-based dissimilarity&quot;,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">     labels = FALSE, hang = 0.04)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain"># &gt;&gt;&gt; 3.2 将 gene 聚类树进行修剪后，把 gene 归到不同的模块里</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">dynamic_modules &lt;- cutreeDynamic(dendro = geneTree,distM = dissTOM,minClusterSize = 50,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                                 deepSplit = 2,pamRespectsDendro=F,minSplitHeight = NULL)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">table(dynamic_modules)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">module_color &lt;- labels2colors(dynamic_modules)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">plotDendroAndColors(geneTree,module_color,dendroLabels = F,hang=0.03,addGuide = T,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                    guideHang = 0.05,groupLabels=&#x27;Dynamic color tree&#x27;,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                    main=&#x27;gene dendrogram and module colors&#x27;)</span><br></span></code></pre><div class="buttonGroup__atx"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_eSgA" aria-hidden="true"><svg viewBox="0 0 24 24" class="copyButtonIcon_y97N"><path fill="currentColor" d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg viewBox="0 0 24 24" class="copyButtonSuccessIcon_LjdS"><path fill="currentColor" d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p><img loading="lazy" alt="image-20230509180335465" src="/BioinfoTech/assets/images/image-20230509180335465-89278be6620ad22dfe97f0fdf1fdc86b.png" width="752" height="514" class="img_ev3q"></p><div class="language-R codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_biex"><pre tabindex="0" class="prism-code language-R codeBlock_bY9V thin-scrollbar"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token plain"># &gt;&gt;&gt; 3.3 计算每个模块的特征基因向量（对每个模块的 gene 表达量降维后，只保留 PC1，即 MEs)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">MEList &lt;- moduleEigengenes(exp,colors = dynamic_modules)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">MEs &lt;- MEList$eigengenes</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain"># &gt;&gt;&gt; 3.4 计算特征基因的相异度，基于相异度对原有模块进行层次聚类</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">dissME &lt;- 1-cor(MEs)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">MEtree &lt;- hclust(as.dist(dissME),method = &#x27;average&#x27;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">plotEigengeneNetworks(MEs, </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                      &quot;Eigengene adjacency heatmap&quot;, </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                      marHeatmap = c(3,4,2,2), </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                      plotDendrograms = FALSE, </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                      xLabelsAngle = 90) </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain"># &gt;&gt;&gt; 3.5 如果模块太多，可以对模块进行合并。设置剪切高度，剪切高度以下的模块进行合并</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">cut_height &lt;- 0.2</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">sizeGrWindow(8,6)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">plot(MEtree,main = &quot;Clustering of module eigengenes&quot;, </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">     xlab = &quot;&quot;, sub = &quot;&quot;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">abline(h=cut_height,col=&#x27;red&#x27;)</span><br></span></code></pre><div class="buttonGroup__atx"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_eSgA" aria-hidden="true"><svg viewBox="0 0 24 24" class="copyButtonIcon_y97N"><path fill="currentColor" d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg viewBox="0 0 24 24" class="copyButtonSuccessIcon_LjdS"><path fill="currentColor" d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p><img loading="lazy" alt="image-20230509180405685" src="/BioinfoTech/assets/images/image-20230509180405685-54d450ca9a557198b402cf9c271d55df.png" width="752" height="418" class="img_ev3q"></p><p>把高度 0.2 以下的模块进行合并，再看看合并前后模块的变化</p><div class="language-R codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_biex"><pre tabindex="0" class="prism-code language-R codeBlock_bY9V thin-scrollbar"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token plain">module_merge &lt;- mergeCloseModules(exp,module_color,cutHeight = cut_height)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">merged_color &lt;- module_merge$colors</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">merged_MEs &lt;- module_merge$newMEs</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">sizeGrWindow(8,6)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">plotDendroAndColors(geneTree,cbind(module_color,merged_color),dendroLabels = F,hang=0.03,addGuide = T,groupLabels = c(&quot;Dynamic Tree Cut&quot;, &quot;Merged dynamic&quot;))</span><br></span></code></pre><div class="buttonGroup__atx"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_eSgA" aria-hidden="true"><svg viewBox="0 0 24 24" class="copyButtonIcon_y97N"><path fill="currentColor" d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg viewBox="0 0 24 24" class="copyButtonSuccessIcon_LjdS"><path fill="currentColor" d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p><img loading="lazy" alt="image-20230509184136135" src="/BioinfoTech/assets/images/image-20230509184136135-8ff1f4eec5acbb47d03b54ad37c3dc13.png" width="752" height="523" class="img_ev3q"></p><div class="language-R codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_biex"><pre tabindex="0" class="prism-code language-R codeBlock_bY9V thin-scrollbar"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token plain"># 4. 计算每个模块与临床信息的相关性、pvalue。找出与临床信息相关性高的模块</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">nGenes &lt;- ncol(exp)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">nSamples &lt;- nrow(exp)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">################################准备临床信息####################################</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">clinical &lt;- clinical[,c(&#x27;bcr_patient_barcode&#x27;,&#x27;gender&#x27;,&#x27;residual_tumor&#x27;)]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">traits &lt;- data.frame(row.names = rownames(exp),ID=rownames(exp))</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">traits$ID &lt;- substr(traits$ID,1,12)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">index &lt;- match(traits$ID,clinical$bcr_patient_barcode)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">traits$gender &lt;- as.numeric(as.factor(clinical[index,&#x27;gender&#x27;]))</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">traits$residual_tumor &lt;- as.numeric(as.factor(clinical[index,&#x27;residual_tumor&#x27;]))</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">traits$type &lt;- ifelse(sapply(str_split(rownames(traits),&#x27;-&#x27;,simplify = T)[,4],</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                                       function(x){as.integer(substr(x,1,2))})&gt;=10,&#x27;normal&#x27;,&#x27;tumor&#x27;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">traits$type &lt;- as.numeric(as.factor(traits$type))</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">traits &lt;- traits[,-1]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">################################准备临床信息####################################</span><br></span></code></pre><div class="buttonGroup__atx"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_eSgA" aria-hidden="true"><svg viewBox="0 0 24 24" class="copyButtonIcon_y97N"><path fill="currentColor" d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg viewBox="0 0 24 24" class="copyButtonSuccessIcon_LjdS"><path fill="currentColor" d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>clinical 是前面下载的 TCGA 临床数据，这里随便取了两个比较关心的临床信息。clinical 的每一行是一个病人的 ID，而表达矩阵每一行是一个样本，一个病人身上可能有多个样本。在把临床信息和表达矩阵一一对应的时候，如果直接把 exp 的行名取前 12 位，就会产生重复值。在做的时候，先准备好样本和病人 ID 的对应关系，再把 clinical 里病人的信息给匹配过来</p><p><img loading="lazy" alt="image-20230509184220708" src="/BioinfoTech/assets/images/image-20230509184220708-f55306ba26b39b92784a368fe80f3217.png" width="752" height="856" class="img_ev3q"></p><p>最终的临床信息是这样，要把字符串变成数值型变量</p><p><img loading="lazy" alt="image-20230509184243649" src="/BioinfoTech/assets/images/image-20230509184243649-98e4c41ecd7f13aebcd8f5831b530599.png" width="752" height="548" class="img_ev3q"></p><div class="language-R codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_biex"><pre tabindex="0" class="prism-code language-R codeBlock_bY9V thin-scrollbar"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token plain">module_trait_cor &lt;- cor(merged_MEs,traits,use = &#x27;p&#x27;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">module_trait_pvalue &lt;- corPvalueStudent(module_trait_cor,nSamples)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">trait_colors &lt;- numbers2colors(traits,signed = T,centered = T)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">a &lt;- paste(signif(module_trait_cor, 2), &quot;\n(&quot;, signif(module_trait_pvalue, 1), &quot;)&quot;, sep = &quot;&quot;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">sizeGrWindow(8,6)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">labeledHeatmap(module_trait_cor,xLabels = colnames(traits),yLabels = colnames(merged_MEs),colorLabels = FALSE, </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">               colors = blueWhiteRed(50),textMatrix = a,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">               setStdMargins = FALSE,cex.text = 0.65,zlim = c(-1,1), </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">               main = paste(&quot;Module-trait relationships&quot;))</span><br></span></code></pre><div class="buttonGroup__atx"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_eSgA" aria-hidden="true"><svg viewBox="0 0 24 24" class="copyButtonIcon_y97N"><path fill="currentColor" d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg viewBox="0 0 24 24" class="copyButtonSuccessIcon_LjdS"><path fill="currentColor" d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>模块与 tumor/normal 之间的相关性并不是很高。就拿粉色模块当做感兴趣的模块吧</p><p><img loading="lazy" alt="image-20230509184310937" src="/BioinfoTech/assets/images/image-20230509184310937-65f0003b4937014f9b63dc249d7711cb.png" width="752" height="552" class="img_ev3q"></p><div class="language-R codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_biex"><pre tabindex="0" class="prism-code language-R codeBlock_bY9V thin-scrollbar"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token plain"># 5. 找出感兴趣的模块内的枢纽 gene</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain"># &gt;&gt;&gt; 5.1 计算 gene 表达量与临床性状的相关性</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">gs &lt;- apply(exp,2,function(x){cor(x,traits$type)})</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain"># &gt;&gt;&gt; 5.2 基于网络计算 gene 与临床性状的相关性</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">ns &lt;- networkScreening(y=traits$type,datME = merged_MEs,datExpr = exp)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain"># &gt;&gt;&gt; 5.3 计算每个 gene 与特征 gene 的相关性</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">kme &lt;- signedKME(datExpr = exp,datME = merged_MEs)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain"># &gt;&gt;&gt; 5.4 根据 gs, ns, kme 找出 hub gene</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">hubgene_filter &lt;- (gs&gt;0.2 &amp; ns$q.Weighted&lt;0.01 &amp; abs(kme$kMEmagenta)&gt;0.4)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">hub_gene &lt;- colnames(exp)[hubgene_filter]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">plotNetworkHeatmap(exp,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                   plotGenes = hub_gene,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                   networkType = &quot;unsigned&quot;,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                   useTOM = TRUE,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                   power=8,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                   main=&quot;unsigned correlations&quot;)</span><br></span></code></pre><div class="buttonGroup__atx"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_eSgA" aria-hidden="true"><svg viewBox="0 0 24 24" class="copyButtonIcon_y97N"><path fill="currentColor" d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg viewBox="0 0 24 24" class="copyButtonSuccessIcon_LjdS"><path fill="currentColor" d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>看下 kme，其中粉色模块里，gene 的相关性并不是很高，这里筛选了&gt;0.4 的，最后找到 25 个 hub gene。如果阈值再设高一点，hub gene 就全卡掉了。这跟没有用所有的 3w 多个 gene 去计算有关，也跟一些参数有关，我没有一个个去调整优化，只是展示一下如何做图和分析思路</p><div class="language-R codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_biex"><pre tabindex="0" class="prism-code language-R codeBlock_bY9V thin-scrollbar"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token plain">hub_gene &lt;- intersect(hub_gene,deg$ID)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">save(hub_gene,file = &#x27;TCGA_WGCNA_hub_gene.Rdata&#x27;)</span><br></span></code></pre><div class="buttonGroup__atx"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_eSgA" aria-hidden="true"><svg viewBox="0 0 24 24" class="copyButtonIcon_y97N"><path fill="currentColor" d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg viewBox="0 0 24 24" class="copyButtonSuccessIcon_LjdS"><path fill="currentColor" d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>最后找出 hub gene 和前面 deg 的交集部分，作为 hub gene，存档</p><h2 class="anchor anchorWithStickyNavbar_LWe7" id="step-4-非负矩阵分解算法">Step 4. 非负矩阵分解算法<a href="#step-4-非负矩阵分解算法" class="hash-link" aria-label="Direct link to Step 4. 非负矩阵分解算法" title="Direct link to Step 4. 非负矩阵分解算法">​</a></h2><h3 class="anchor anchorWithStickyNavbar_LWe7" id="41-nmf-算法对样本分类">4.1 NMF 算法对样本分类<a href="#41-nmf-算法对样本分类" class="hash-link" aria-label="Direct link to 4.1 NMF 算法对样本分类" title="Direct link to 4.1 NMF 算法对样本分类">​</a></h3><div class="language-R codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_biex"><pre tabindex="0" class="prism-code language-R codeBlock_bY9V thin-scrollbar"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token plain">library(NMF)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">library(survival)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">library(stringr)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">library(data.table)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">library(survminer)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">library(dplyr)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">library(AnnoProbe)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">library(CIBERSORT)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">library(tidyr)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">library(tibble)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">library(RColorBrewer)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">library(ggalluvial)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">library(ggplot2)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">options(stringsAsFactors = F)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">rm(list = ls())</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">gc()</span><br></span></code></pre><div class="buttonGroup__atx"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_eSgA" aria-hidden="true"><svg viewBox="0 0 24 24" class="copyButtonIcon_y97N"><path fill="currentColor" d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg viewBox="0 0 24 24" class="copyButtonSuccessIcon_LjdS"><path fill="currentColor" d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><div class="language-R codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_biex"><pre tabindex="0" class="prism-code language-R codeBlock_bY9V thin-scrollbar"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token plain">load(&#x27;TCGA_LUAD_TPM_exp.Rdata&#x27;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">load(&#x27;TCGA_DEGS.Rdata&#x27;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">load(&#x27;TCGA_WGCNA_hub_gene.Rdata&#x27;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">dim(exp)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain"># 对表达矩阵取 log2</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">exp &lt;- log2(exp+1)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">ranks &lt;- seq(2,10)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">hubgene_exp &lt;- exp[hub_gene,]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">estimate &lt;- nmf(hubgene_exp,2:10,nrun=50,method = &#x27;brunet&#x27;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">plot(estimate)</span><br></span></code></pre><div class="buttonGroup__atx"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_eSgA" aria-hidden="true"><svg viewBox="0 0 24 24" class="copyButtonIcon_y97N"><path fill="currentColor" d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg viewBox="0 0 24 24" class="copyButtonSuccessIcon_LjdS"><path fill="currentColor" d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>注意 hubgene<!-- -->_<!-- -->exp 里不能有负数，不然会报错，rank 为 2 时，波动最大</p><p><img loading="lazy" alt="image-20230509184405897" src="/BioinfoTech/assets/images/image-20230509184405897-034a0d5f9b5bcb78eafaf7672842dd10.png" width="752" height="535" class="img_ev3q"></p><div class="language-R codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_biex"><pre tabindex="0" class="prism-code language-R codeBlock_bY9V thin-scrollbar"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token plain">rank &lt;- 2</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">nmf.rank4 &lt;- nmf(hubgene_exp,rank,nrun=50,method = &#x27;brunet&#x27;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">index &lt;- extractFeatures(nmf.rank4,&#x27;max&#x27;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">index &lt;- unlist(index)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">nmf.exp &lt;- hubgene_exp[index,]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">nmf.exp &lt;- na.omit(nmf.exp)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">group &lt;- predict(nmf.rank4)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">jco &lt;- c(&quot;#2874C5&quot;,&quot;#EABF00&quot;,&quot;#C6524A&quot;,&quot;#868686&quot;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">consensusmap(nmf.rank4,labRow=NA,labCol=NA,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">             annCol = data.frame(&quot;cluster&quot;=group[colnames(nmf.exp)]),</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">             annColors = list(cluster=c(&quot;1&quot;=jco[1],&quot;2&quot;=jco[2],&quot;3&quot;=jco[3],&quot;4&quot;=jco[4])))</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">cluster_df &lt;- data.frame(row.names = colnames(exp),ID=substr(colnames(exp),1,12),cluster=group)</span><br></span></code></pre><div class="buttonGroup__atx"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_eSgA" aria-hidden="true"><svg viewBox="0 0 24 24" class="copyButtonIcon_y97N"><path fill="currentColor" d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg viewBox="0 0 24 24" class="copyButtonSuccessIcon_LjdS"><path fill="currentColor" d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>NMF 算法把所有样本分成两类，顺便看下最后的 clueter<!-- -->_<!-- -->df<!-- -->![图片]<!-- -->(data:image/svg+xml,%3C%3Fxml version=&#x27;1.0&#x27; encoding=&#x27;UTF-8&#x27;%3F%3E%3Csvg width=&#x27;1px&#x27; height=&#x27;1px&#x27; viewBox=&#x27;0 0 1 1&#x27; version=&#x27;1.1&#x27; xmlns=&#x27;<a href="http://www.w3.org/2000/svg&#x27;" target="_blank" rel="noopener noreferrer">http://www.w3.org/2000/svg&#x27;</a> xmlns:xlink=&#x27;<a href="http://www.w3.org/1999/xlink&#x27;%3E%3Ctitle%3E%3C/title%3E%3Cg" target="_blank" rel="noopener noreferrer">http://www.w3.org/1999/xlink&#x27;%3E%3Ctitle%3E%3C/title%3E%3Cg</a> stroke=&#x27;none&#x27; stroke-width=&#x27;1&#x27; fill=&#x27;none&#x27; fill-rule=&#x27;evenodd&#x27; fill-opacity=&#x27;0&#x27;%3E%3Cg transform=&#x27;translate(-249.000000, -126.000000)&#x27; fill=&#x27;%23FFFFFF&#x27;%3E%3Crect x=&#x27;249&#x27; y=&#x27;126&#x27; width=&#x27;1&#x27; height=&#x27;1&#x27;%3E%3C/rect%3E%3C/g%3E%3C/g%3E%3C/svg%3E)</p><p><img loading="lazy" alt="image-20230509184432644" src="/BioinfoTech/assets/images/image-20230509184432644-eff505881be9cbcf479881ece10c2b43.png" width="752" height="1141" class="img_ev3q"></p><h3 class="anchor anchorWithStickyNavbar_LWe7" id="42-生存分析">4.2 生存分析<a href="#42-生存分析" class="hash-link" aria-label="Direct link to 4.2 生存分析" title="Direct link to 4.2 生存分析">​</a></h3><div class="language-R codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_biex"><pre tabindex="0" class="prism-code language-R codeBlock_bY9V thin-scrollbar"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token plain">load(&#x27;TCGA_LUAD_TPM_exp.Rdata&#x27;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">load(&#x27;TCGA_LUAD_Clinical.Rdata&#x27;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain"># 1. 取 log（TPM）</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">for (i in 1:ncol(exp)){</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  exp[,i] &lt;- log2(exp[,i]+0.0001)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">}</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain"># 2. 以病人为中心，去重</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">exp &lt;- as.data.frame(exp)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">exp &lt;- exp[,sort(colnames(exp))]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">index &lt;- !duplicated(substr(colnames(exp),1,12))</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">exp &lt;- exp[,index]</span><br></span></code></pre><div class="buttonGroup__atx"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_eSgA" aria-hidden="true"><svg viewBox="0 0 24 24" class="copyButtonIcon_y97N"><path fill="currentColor" d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg viewBox="0 0 24 24" class="copyButtonSuccessIcon_LjdS"><path fill="currentColor" d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><div class="language-R codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_biex"><pre tabindex="0" class="prism-code language-R codeBlock_bY9V thin-scrollbar"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token plain"># 3. 读取从 xena 下载的生存数据</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">sur &lt;- fread(&#x27;TCGA-LUAD.survival.tsv&#x27;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">sur &lt;- as.data.frame(sur)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">patient_id &lt;- substr(colnames(exp),1,12)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">clin &lt;- data.frame(row.names = patient_id,ID=patient_id)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">index &lt;- match(clin$ID,sur[,&#x27;_PATIENT&#x27;])</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">clin &lt;- cbind(clin,sur[index,c(&#x27;OS&#x27;,&#x27;OS.time&#x27;)])</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">clin &lt;- na.omit(clin)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">index2 &lt;- match(clin$ID,clinical$bcr_patient_barcode)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">clin &lt;- cbind(clin,clinical[index2,&#x27;gender&#x27;])</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">colnames(clin) &lt;- c(colnames(clin)[1:ncol(clin)-1],&#x27;gender&#x27;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">clin$gender &lt;- as.numeric(as.factor(clin$gender))</span><br></span></code></pre><div class="buttonGroup__atx"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_eSgA" aria-hidden="true"><svg viewBox="0 0 24 24" class="copyButtonIcon_y97N"><path fill="currentColor" d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg viewBox="0 0 24 24" class="copyButtonSuccessIcon_LjdS"><path fill="currentColor" d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p><img loading="lazy" alt="image-20230509184553116" src="/BioinfoTech/assets/images/image-20230509184553116-7e2642d1b1039608fcf6da84e7584f79.png" width="757" height="772" class="img_ev3q"></p><p><img loading="lazy" alt="image-20230509184613334" src="/BioinfoTech/assets/images/image-20230509184613334-d2a4823290a09a0871e7942ab593abc3.png" width="757" height="730" class="img_ev3q"></p><p>下载下来的生存数据长这样</p><p><img loading="lazy" alt="image-20230509184731529" src="/BioinfoTech/assets/images/image-20230509184731529-1ff4a12b475a511658b6a23ec340ec1d.png" width="757" height="383" class="img_ev3q"></p><p>之前从 TCGA 下载的临床数据长这样</p><p><img loading="lazy" alt="image-20230509184831481" src="/BioinfoTech/assets/images/image-20230509184831481-f0bb5d74b31d53113229d00778cb8be6.png" width="757" height="278" class="img_ev3q"></p><p>合并后的 clin 是这样，所有变量类型都要转成数值型</p><p><img loading="lazy" alt="image-20230509184848098" src="/BioinfoTech/assets/images/image-20230509184848098-ff586e81f0aae9e8c65943deda2fbad4.png" width="757" height="462" class="img_ev3q"></p><div class="language-R codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_biex"><pre tabindex="0" class="prism-code language-R codeBlock_bY9V thin-scrollbar"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token plain"># 合并 NMF 结果</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">cluster_type_group &lt;- ifelse(sapply(str_split(rownames(cluster_df),&#x27;-&#x27;,simplify = T)[,4],</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                                    function(x){as.integer(substr(x,1,2))}</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                                    )&gt;=10,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                            &#x27;normal&#x27;,&#x27;tumor&#x27;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">cluster_df &lt;- cluster_df[which(cluster_type_group==&#x27;tumor&#x27;),]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">clin &lt;- left_join(clin,cluster_df,&#x27;ID&#x27;,multiple=&#x27;first&#x27;)</span><br></span></code></pre><div class="buttonGroup__atx"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_eSgA" aria-hidden="true"><svg viewBox="0 0 24 24" class="copyButtonIcon_y97N"><path fill="currentColor" d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg viewBox="0 0 24 24" class="copyButtonSuccessIcon_LjdS"><path fill="currentColor" d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>再把 NMF 分类的结果添加进来，就是 cluster 那一列</p><p><img loading="lazy" alt="image-20230509184928906" src="/BioinfoTech/assets/images/image-20230509184928906-1ffbe0011d9e37d9075ca5263b3a4359.png" width="748" height="410" class="img_ev3q"></p><div class="language-R codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_biex"><pre tabindex="0" class="prism-code language-R codeBlock_bY9V thin-scrollbar"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token plain"># 7. 过滤生存数据，不要&lt;30 天的，再把天数转换成月</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">clin &lt;- clin[clin$OS.time&gt;30,]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">clin$OS.time &lt;- clin$OS.time / 30</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain"># 8. exp 的列名和临床信息的行名一一对应。NMF 将样本分成若干个亚型，画出亚型的生存曲线</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">colnames(exp) &lt;- substr(colnames(exp),1,12)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">index &lt;- match(clin$ID,colnames(exp))</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">exp &lt;- exp[,index]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">sfit &lt;- survfit(Surv(OS.time,OS)~cluster,data=clin)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">ggsurvplot(sfit,pval = T)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">save(clin,file = &#x27;clin.Rdata&#x27;)</span><br></span></code></pre><div class="buttonGroup__atx"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_eSgA" aria-hidden="true"><svg viewBox="0 0 24 24" class="copyButtonIcon_y97N"><path fill="currentColor" d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg viewBox="0 0 24 24" class="copyButtonSuccessIcon_LjdS"><path fill="currentColor" d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p><img loading="lazy" alt="image-20230509184951924" src="/BioinfoTech/assets/images/image-20230509184951924-a578e93dc8e247c366c9a791b9017d22.png" width="748" height="516" class="img_ev3q"></p><h3 class="anchor anchorWithStickyNavbar_LWe7" id="43-免疫浸润">4.3 免疫浸润<a href="#43-免疫浸润" class="hash-link" aria-label="Direct link to 4.3 免疫浸润" title="Direct link to 4.3 免疫浸润">​</a></h3><div class="language-R codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_biex"><pre tabindex="0" class="prism-code language-R codeBlock_bY9V thin-scrollbar"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token plain">my_pallet &lt;- colorRampPalette(brewer.pal(8,&#x27;Set1&#x27;))</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">lm22f = system.file(&quot;extdata&quot;, &quot;LM22.txt&quot;, package = &quot;CIBERSORT&quot;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">cluster1 &lt;- exp[,rownames(cluster_df[cluster_df$cluster==1,])]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">cluster2 &lt;- exp[,rownames(cluster_df[cluster_df$cluster==2,])]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">TME.result1 &lt;- cibersort(lm22f,as.matrix(cluster1),perm = 100,QN=T) # 耗时很久</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">TME.result2 &lt;- cibersort(lm22f,as.matrix(cluster2),perm = 100,QN=T)</span><br></span></code></pre><div class="buttonGroup__atx"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_eSgA" aria-hidden="true"><svg viewBox="0 0 24 24" class="copyButtonIcon_y97N"><path fill="currentColor" d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg viewBox="0 0 24 24" class="copyButtonSuccessIcon_LjdS"><path fill="currentColor" d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>原文中免疫浸润用 MCPcounter 做的，这里我用 CIBERSORT 做，运行完的结果就是各种免疫细胞在每个样本中的含量</p><div class="language-R codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_biex"><pre tabindex="0" class="prism-code language-R codeBlock_bY9V thin-scrollbar"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token plain">result1 &lt;- TME.result1[,-c(23:25)] %&gt;% as.data.frame() %&gt;% </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  rownames_to_column(&#x27;Sample&#x27;) %&gt;% gather(key = cell_type,value = proportion,-Sample)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">result2 &lt;- TME.result2[,-c(23:25)] %&gt;% as.data.frame() %&gt;% </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  rownames_to_column(&#x27;Sample&#x27;) %&gt;% gather(key = cell_type,value = proportion,-Sample)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">result1$cluster &lt;- rep(&#x27;cluster1&#x27;, times=nrow(result1))</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">result2$cluster &lt;- rep(&#x27;cluster2&#x27;, times=nrow(result2))</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">dat &lt;- rbind(result1,result2)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">ggplot(dat,aes(axis1 = cluster, axis2 = cell_type))+</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  geom_alluvium(aes(fill = as.factor(cell_type)),width = 2/5, discern = FALSE)+</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  geom_stratum(width = 2/5, discern = FALSE)+</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  geom_text(stat = &quot;stratum&quot;, discern = FALSE,aes(label = after_stat(stratum)))+</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  theme(legend.position = &quot;none&quot;,#去除刻度线和背景颜色</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">        panel.background = element_blank(),</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">        axis.ticks = element_blank(),</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">        axis.text.y = element_blank(),</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">        axis.text.x = element_text(size =15,colour = &quot;black&quot;),#坐标轴名</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">        axis.title = element_blank()) +</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  scale_x_discrete(position = &quot;top&quot;)</span><br></span></code></pre><div class="buttonGroup__atx"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_eSgA" aria-hidden="true"><svg viewBox="0 0 24 24" class="copyButtonIcon_y97N"><path fill="currentColor" d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg viewBox="0 0 24 24" class="copyButtonSuccessIcon_LjdS"><path fill="currentColor" d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>把上面的透视表转换成长表，添加上 cluster 的信息，再把 cluster1 和 cluster2 对应的长表合并起来，就可以画图了</p><p><img loading="lazy" alt="image-20230509185053953" src="/BioinfoTech/assets/images/image-20230509185053953-8987ba1c815e0c0a9d33ccd0abd2e75b.png" width="748" height="756" class="img_ev3q"></p><div class="language-R codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_biex"><pre tabindex="0" class="prism-code language-R codeBlock_bY9V thin-scrollbar"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token plain">cell &lt;- &#x27;Plasma cells&#x27;</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">ggviolin(dat[dat$cell_type==cell,],x=&#x27;cluster&#x27;,y=&#x27;proportion&#x27;,fill = &#x27;cluster&#x27;,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">         add = &#x27;boxplot&#x27;,add.params = list(fill = &#x27;white&#x27;,width=0.1,linetype=1),title = cell)+</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  stat_compare_means(method = &#x27;t.test&#x27;,label = &#x27;p.format&#x27;,aes(label = &#x27;p.format&#x27;),</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                              label.x.npc = &#x27;center&#x27;,label.y = 0.5,size=5)</span><br></span></code></pre><div class="buttonGroup__atx"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_eSgA" aria-hidden="true"><svg viewBox="0 0 24 24" class="copyButtonIcon_y97N"><path fill="currentColor" d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg viewBox="0 0 24 24" class="copyButtonSuccessIcon_LjdS"><path fill="currentColor" d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p><img loading="lazy" alt="image-20230509185114933" src="/BioinfoTech/assets/images/image-20230509185114933-ea1d84302de2426872586568cfd5f77a.png" width="748" height="535" class="img_ev3q"></p><h3 class="anchor anchorWithStickyNavbar_LWe7" id="44-批量单因素-cox-回归">4.4 批量单因素 Cox 回归<a href="#44-批量单因素-cox-回归" class="hash-link" aria-label="Direct link to 4.4 批量单因素 Cox 回归" title="Direct link to 4.4 批量单因素 Cox 回归">​</a></h3><div class="language-R codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_biex"><pre tabindex="0" class="prism-code language-R codeBlock_bY9V thin-scrollbar"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token plain">load(&#x27;TCGA_WGCNA_hub_gene.Rdata&#x27;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">load(&#x27;TCGA_LUAD_TPM_exp.Rdata&#x27;) </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">load(&#x27;clin.Rdata&#x27;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">hubgene_exp &lt;- exp[hub_gene,]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">hubgene_exp &lt;- as.data.frame(t(hubgene_exp))</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">hubgene_exp$ID &lt;- substr(rownames(hubgene_exp),1,12)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">clin &lt;- left_join(clin,hubgene_exp,by=&#x27;ID&#x27;)</span><br></span></code></pre><div class="buttonGroup__atx"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_eSgA" aria-hidden="true"><svg viewBox="0 0 24 24" class="copyButtonIcon_y97N"><path fill="currentColor" d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg viewBox="0 0 24 24" class="copyButtonSuccessIcon_LjdS"><path fill="currentColor" d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>把 WGCNA 中找到的 hub gene 挑出来，拿到他们在所有样本中的表达值。把表达值和生存信息合并到一起，最后的 clin 长这样。在合并之前要注意，删掉 normal 样本，我这里没有做。<!-- -->![图片]<!-- -->(data:image/svg+xml,%3C%3Fxml version=&#x27;1.0&#x27; encoding=&#x27;UTF-8&#x27;%3F%3E%3Csvg width=&#x27;1px&#x27; height=&#x27;1px&#x27; viewBox=&#x27;0 0 1 1&#x27; version=&#x27;1.1&#x27; xmlns=&#x27;<a href="http://www.w3.org/2000/svg&#x27;" target="_blank" rel="noopener noreferrer">http://www.w3.org/2000/svg&#x27;</a> xmlns:xlink=&#x27;<a href="http://www.w3.org/1999/xlink&#x27;%3E%3Ctitle%3E%3C/title%3E%3Cg" target="_blank" rel="noopener noreferrer">http://www.w3.org/1999/xlink&#x27;%3E%3Ctitle%3E%3C/title%3E%3Cg</a> stroke=&#x27;none&#x27; stroke-width=&#x27;1&#x27; fill=&#x27;none&#x27; fill-rule=&#x27;evenodd&#x27; fill-opacity=&#x27;0&#x27;%3E%3Cg transform=&#x27;translate(-249.000000, -126.000000)&#x27; fill=&#x27;%23FFFFFF&#x27;%3E%3Crect x=&#x27;249&#x27; y=&#x27;126&#x27; width=&#x27;1&#x27; height=&#x27;1&#x27;%3E%3C/rect%3E%3C/g%3E%3C/g%3E%3C/svg%3E)</p><div class="language-R codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_biex"><pre tabindex="0" class="prism-code language-R codeBlock_bY9V thin-scrollbar"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token plain">res.cox &lt;- coxph(Surv(OS.time,OS)~SAPCD2,data=clin)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">sumres.cox &lt;- summary(res.cox)</span><br></span></code></pre><div class="buttonGroup__atx"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_eSgA" aria-hidden="true"><svg viewBox="0 0 24 24" class="copyButtonIcon_y97N"><path fill="currentColor" d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg viewBox="0 0 24 24" class="copyButtonSuccessIcon_LjdS"><path fill="currentColor" d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>挑一个基因 SAPCD2 做单因素 cox 回归，最后需要的信息都藏在 sumres.cox 里</p><div class="language-R codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_biex"><pre tabindex="0" class="prism-code language-R codeBlock_bY9V thin-scrollbar"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token plain"># 批量单因素 cox 回归</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">gene &lt;- c()</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">p_value &lt;- c()</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">HR &lt;- c()</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">lower95 &lt;- c()</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">upper95 &lt;- c()</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">for (i in 6:ncol(clin)){</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  res &lt;- coxph(Surv(OS.time,OS)~clin[,i],data=clin)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  sum_res &lt;- summary(res)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  p &lt;- sum_res$coefficients[,&#x27;Pr(&gt;|z|)&#x27;]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  if (p&lt;0.05){</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    gene &lt;- c(gene,colnames(clin)[i])</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    p_value &lt;- c(p_value,p)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    HR &lt;- c(HR,sum_res$conf.int[,&#x27;exp(coef)&#x27;])</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    lower95 &lt;- c(lower95,sum_res$conf.int[,&#x27;lower .95&#x27;])</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    upper95 &lt;- c(upper95,sum_res$conf.int[,&#x27;upper .95&#x27;])</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  }</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">}</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">cox_df &lt;- data.frame(row.names = gene,pvalue=p_value,HR=HR,lower.95=lower95,upper.95=upper95)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">lasso_input_gene &lt;- rownames(cox_df[cox_df$pvalue&lt;0.01,])</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">save(lasso_input_gene,file = &#x27;Lasso_input_gene.Rdata&#x27;)</span><br></span></code></pre><div class="buttonGroup__atx"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_eSgA" aria-hidden="true"><svg viewBox="0 0 24 24" class="copyButtonIcon_y97N"><path fill="currentColor" d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg viewBox="0 0 24 24" class="copyButtonSuccessIcon_LjdS"><path fill="currentColor" d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>批量单因素 cox 回归得到每个 gene 的 pvalue，挑出 pvalue&lt;0.01 的再进行后续筛选</p><h2 class="anchor anchorWithStickyNavbar_LWe7" id="step-5-预后模型构建和验证">Step 5. 预后模型构建和验证<a href="#step-5-预后模型构建和验证" class="hash-link" aria-label="Direct link to Step 5. 预后模型构建和验证" title="Direct link to Step 5. 预后模型构建和验证">​</a></h2><div class="language-R codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_biex"><pre tabindex="0" class="prism-code language-R codeBlock_bY9V thin-scrollbar"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token plain">load(&#x27;Lasso_input_gene.Rdata&#x27;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">load(&#x27;TCGA_LUAD_TPM_exp.Rdata&#x27;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">load(&#x27;clin.Rdata&#x27;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain"># 对表达矩阵取 log2</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">exp &lt;- log2(exp+0.00001)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">exp &lt;- as.data.frame(exp)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain"># Lasso 回归进一步筛选 gene</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">exp2 &lt;- exp[lasso_input_gene,]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">exp2 &lt;- as.data.frame(t(exp2))</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">exp2$ID &lt;- substr(rownames(exp2),1,12)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">exp2 &lt;- exp2[!duplicated(exp2$ID),]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">clin &lt;- clin[,1:3]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">exp3 &lt;- left_join(clin,exp2,by=&#x27;ID&#x27;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">rownames(exp3) &lt;- exp3$ID</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">exp3 &lt;- exp3[,-1]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">exp3 &lt;- na.omit(exp3)</span><br></span></code></pre><div class="buttonGroup__atx"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_eSgA" aria-hidden="true"><svg viewBox="0 0 24 24" class="copyButtonIcon_y97N"><path fill="currentColor" d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg viewBox="0 0 24 24" class="copyButtonSuccessIcon_LjdS"><path fill="currentColor" d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>exp3 是生存数据和上一步批量单因素 Cox 回归筛选出来的基因的表达量合并起来的结果！<!-- -->[图片]<!-- -->(data:image/svg+xml,%3C%3Fxml version=&#x27;1.0&#x27; encoding=&#x27;UTF-8&#x27;%3F%3E%3Csvg width=&#x27;1px&#x27; height=&#x27;1px&#x27; viewBox=&#x27;0 0 1 1&#x27; version=&#x27;1.1&#x27; xmlns=&#x27;<a href="http://www.w3.org/2000/svg&#x27;" target="_blank" rel="noopener noreferrer">http://www.w3.org/2000/svg&#x27;</a> xmlns:xlink=&#x27;<a href="http://www.w3.org/1999/xlink&#x27;%3E%3Ctitle%3E%3C/title%3E%3Cg" target="_blank" rel="noopener noreferrer">http://www.w3.org/1999/xlink&#x27;%3E%3Ctitle%3E%3C/title%3E%3Cg</a> stroke=&#x27;none&#x27; stroke-width=&#x27;1&#x27; fill=&#x27;none&#x27; fill-rule=&#x27;evenodd&#x27; fill-opacity=&#x27;0&#x27;%3E%3Cg transform=&#x27;translate(-249.000000, -126.000000)&#x27; fill=&#x27;%23FFFFFF&#x27;%3E%3Crect x=&#x27;249&#x27; y=&#x27;126&#x27; width=&#x27;1&#x27; height=&#x27;1&#x27;%3E%3C/rect%3E%3C/g%3E%3C/g%3E%3C/svg%3E)</p><div class="language-R codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_biex"><pre tabindex="0" class="prism-code language-R codeBlock_bY9V thin-scrollbar"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token plain">x &lt;- as.matrix(exp3[,3:(ncol(exp3)-1)])</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">y &lt;- Surv(exp3$OS.time,exp3$OS)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">fit1 &lt;- glmnet(x,y,alpha = 1,family = &#x27;cox&#x27;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">plot(fit1)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">cv_fit &lt;- cv.glmnet(x,y,alpha = 1,nfolds = 10,family=&quot;cox&quot;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">plot(cv_fit)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">best_lambda &lt;- cv_fit$lambda.min</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">fit2 &lt;- glmnet(x,y,lambda = best_lambda,alpha = 1,family = &#x27;cox&#x27;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">lasso_filter_gene &lt;- names(fit2$beta[fit2$beta[,1]!=0,1])</span><br></span></code></pre><div class="buttonGroup__atx"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_eSgA" aria-hidden="true"><svg viewBox="0 0 24 24" class="copyButtonIcon_y97N"><path fill="currentColor" d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg viewBox="0 0 24 24" class="copyButtonSuccessIcon_LjdS"><path fill="currentColor" d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>由于之前筛选出来的 gene 效果不怎么好， 经过 lasso 过滤后，只剩 2 个 gene 了，就用这两个来做多因素 cox 回归</p><div class="language-R codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_biex"><pre tabindex="0" class="prism-code language-R codeBlock_bY9V thin-scrollbar"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token plain"># 多因素 cox 回归构建预后模型</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">exp4 &lt;- exp3 %&gt;% dplyr::select(OS,OS.time,lasso_filter_gene)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">multicox &lt;- coxph(Surv(OS.time,OS)~.,data = exp4)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">sum_multicox &lt;- summary(multicox)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">riskScore &lt;- predict(multicox,type = &#x27;risk&#x27;,newdata = exp4)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">riskScore &lt;- as.data.frame(riskScore)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">riskScore$ID &lt;- rownames(riskScore)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">riskScore$risk &lt;- ifelse(riskScore$riskScore&gt;median(riskScore$riskScore),&#x27;high&#x27;,&#x27;low&#x27;)</span><br></span></code></pre><div class="buttonGroup__atx"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_eSgA" aria-hidden="true"><svg viewBox="0 0 24 24" class="copyButtonIcon_y97N"><path fill="currentColor" d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg viewBox="0 0 24 24" class="copyButtonSuccessIcon_LjdS"><path fill="currentColor" d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>做完多因素回归后，计算每个样本的 riskscore，用 riskScore 把样本分成两类</p><div class="language-R codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_biex"><pre tabindex="0" class="prism-code language-R codeBlock_bY9V thin-scrollbar"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token plain"># KM plot</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">km_data &lt;- left_join(riskScore,clin,by=&#x27;ID&#x27;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">fit &lt;- survfit(Surv(OS.time,OS)~risk,data=km_data)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">ggsurvplot(fit,data = km_data,pval = T,risk.table = T,surv.median.line = &#x27;hv&#x27;,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                       legend.title = &#x27;RiskScore&#x27;,title = &#x27;Overall survival&#x27;,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                       ylab=&#x27;Cummulative survival&#x27;,xlab=&#x27;Time(Days)&#x27;)</span><br></span></code></pre><div class="buttonGroup__atx"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_eSgA" aria-hidden="true"><svg viewBox="0 0 24 24" class="copyButtonIcon_y97N"><path fill="currentColor" d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg viewBox="0 0 24 24" class="copyButtonSuccessIcon_LjdS"><path fill="currentColor" d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p><img loading="lazy" alt="image-20230509185217444" src="/BioinfoTech/assets/images/image-20230509185217444-ba871e9d1e7e4c0d7a1d2191cc5baf06.png" width="748" height="532" class="img_ev3q"></p><div class="language-R codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_biex"><pre tabindex="0" class="prism-code language-R codeBlock_bY9V thin-scrollbar"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token plain"># 绘制 ROC 曲线</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">roc_data &lt;- data.frame(x = roc$FP[,1],</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                       y = roc$TP[,1],</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                       time = roc$times[1])</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">for (i in 2:length(roc$times)) {</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  temp &lt;- data.frame(x = roc$FP[,i],</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                     y = roc$TP[,i],</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                     time = roc$times[i])</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  roc_data &lt;- rbind(roc_data, temp)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">}</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">ggplot(roc_data, aes(x = x, y = y, color = as.factor(time))) + </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  geom_line() + </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  scale_color_manual(values = c(&quot;gray&quot;, &quot;blue&quot;, &quot;black&quot;)) +</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  geom_abline(slope = 1, intercept = 0, linetype = &quot;dashed&quot;) +</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  xlab(&quot;1 - Specificity&quot;) + </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  ylab(&quot;Sensitivity&quot;) + </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  ggtitle(&quot;ROC Curve&quot;) + </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  theme_minimal()</span><br></span></code></pre><div class="buttonGroup__atx"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_eSgA" aria-hidden="true"><svg viewBox="0 0 24 24" class="copyButtonIcon_y97N"><path fill="currentColor" d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg viewBox="0 0 24 24" class="copyButtonSuccessIcon_LjdS"><path fill="currentColor" d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p><img loading="lazy" alt="image-20230509185244243" src="/BioinfoTech/assets/images/image-20230509185244243-7b2b9ede72c60561af609c3710216301.png" width="748" height="533" class="img_ev3q"></p><div class="language-R codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_biex"><pre tabindex="0" class="prism-code language-R codeBlock_bY9V thin-scrollbar"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token plain"># 绘制风险图</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">ggrisk(multicox)</span><br></span></code></pre><div class="buttonGroup__atx"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_eSgA" aria-hidden="true"><svg viewBox="0 0 24 24" class="copyButtonIcon_y97N"><path fill="currentColor" d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg viewBox="0 0 24 24" class="copyButtonSuccessIcon_LjdS"><path fill="currentColor" d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>!<img loading="lazy" alt="image-20230509185309625" src="/BioinfoTech/assets/images/image-20230509185309625-4939cc341e6d9f98aeb89bae8574e497.png" width="748" height="544" class="img_ev3q"></p><div class="language-R codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_biex"><pre tabindex="0" class="prism-code language-R codeBlock_bY9V thin-scrollbar"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token plain"># risk score 与临床病理特征的相关性</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">load(&#x27;TCGA_LUAD_Clinical.Rdata&#x27;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">clinical_df &lt;- clinical[,c(&#x27;bcr_patient_barcode&#x27;,&#x27;gender&#x27;,&#x27;stage_event_pathologic_stage&#x27;)]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">clinical_df &lt;- merge(km_data,clinical_df,by.x = &#x27;ID&#x27;,by.y = &#x27;bcr_patient_barcode&#x27;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">clinical_df &lt;- clinical_df[!duplicated(clinical_df$ID),]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">clinical_df$stage_event_pathologic_stage &lt;- case_when(clinical_df$stage_event_pathologic_stage==&#x27;Stage IA&#x27;~&#x27;Stage I&#x27;,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                                                      clinical_df$stage_event_pathologic_stage==&#x27;Stage IB&#x27;~&#x27;Stage I&#x27;,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                                                      clinical_df$stage_event_pathologic_stage==&#x27;Stage IIA&#x27;~&#x27;Stage II&#x27;,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                                                      clinical_df$stage_event_pathologic_stage==&#x27;Stage IIB&#x27;~&#x27;Stage II&#x27;,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                                                      clinical_df$stage_event_pathologic_stage==&#x27;Stage IIIA&#x27;~&#x27;Stage III&#x27;,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                                                      clinical_df$stage_event_pathologic_stage==&#x27;Stage IIIB&#x27;~&#x27;Stage III&#x27;,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                                                      )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">ggviolin(clinical_df,x=&#x27;gender&#x27;,y=&#x27;riskScore&#x27;,fill = &#x27;gender&#x27;,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">         add = &#x27;boxplot&#x27;,add.params = list(fill = &#x27;white&#x27;,width=0.1,linetype=1),)+</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  stat_compare_means(method = &#x27;wilcox.test&#x27;,label = &#x27;p.format&#x27;,aes(label = &#x27;p.format&#x27;),</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                     label.x.npc = &#x27;center&#x27;,label.y = 20,size=5)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">ggviolin(clinical_df,x=&#x27;stage_event_pathologic_stage&#x27;,y=&#x27;riskScore&#x27;,fill = &#x27;stage_event_pathologic_stage&#x27;,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">         add = &#x27;boxplot&#x27;,add.params = list(fill = &#x27;white&#x27;,width=0.1,linetype=1),)+</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  stat_compare_means(method = &#x27;wilcox.test&#x27;,label = &#x27;p.signif&#x27;,aes(label = &#x27;p.format&#x27;),</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                     label.x = 1,label.y = 20,size=5)</span><br></span></code></pre><div class="buttonGroup__atx"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_eSgA" aria-hidden="true"><svg viewBox="0 0 24 24" class="copyButtonIcon_y97N"><path fill="currentColor" d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg viewBox="0 0 24 24" class="copyButtonSuccessIcon_LjdS"><path fill="currentColor" d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p><img loading="lazy" alt="image-20230509185329453" src="/BioinfoTech/assets/images/image-20230509185329453-223ced0598cb4d6c288b7991b6a3b738.png" width="748" height="1053" class="img_ev3q"></p><h2 class="anchor anchorWithStickyNavbar_LWe7" id="step-6-riskscore-和临床信息的相关性">Step 6. riskScore 和临床信息的相关性<a href="#step-6-riskscore-和临床信息的相关性" class="hash-link" aria-label="Direct link to Step 6. riskScore 和临床信息的相关性" title="Direct link to Step 6. riskScore 和临床信息的相关性">​</a></h2><div class="language-R codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_biex"><pre tabindex="0" class="prism-code language-R codeBlock_bY9V thin-scrollbar"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token plain"># 森林图</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain"># 批量单因素回归</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">df_forest &lt;- clinical_df</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">df_forest$gender &lt;- as.numeric(as.factor(df_forest$gender))</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">df_forest$stage_event_pathologic_stage &lt;- as.numeric(as.factor(df_forest$stage_event_pathologic_stage))</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">gene &lt;- c()</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">p_value &lt;- c()</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">HR &lt;- c()</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">lower95 &lt;- c()</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">upper95 &lt;- c()</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">means &lt;- c()</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">for (i in c(&#x27;riskScore&#x27;,&#x27;gender&#x27;,&#x27;stage_event_pathologic_stage&#x27;)){</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  res &lt;- coxph(Surv(OS.time,OS)~df_forest[,i],data=df_forest)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  sum_res &lt;- summary(res)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  p &lt;- sum_res$coefficients[,&#x27;Pr(&gt;|z|)&#x27;]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  gene &lt;- c(gene,i)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  p_value &lt;- c(p_value,p)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  HR &lt;- c(HR,sum_res$conf.int[,&#x27;exp(coef)&#x27;])</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  lower95 &lt;- c(lower95,sum_res$conf.int[,&#x27;lower .95&#x27;])</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  upper95 &lt;- c(upper95,sum_res$conf.int[,&#x27;upper .95&#x27;])</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  means &lt;- c(means,res$means)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">}</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">cox_df &lt;- data.frame(row.names = gene,pvalue=p_value,HR=HR,lower.95=lower95,upper.95=upper95,means=means)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">cox_df$pvalue &lt;- round(cox_df$pvalue,3)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">cox_df$HR &lt;- round(cox_df$HR,3)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">cox_df &lt;- rownames_to_column(cox_df,var = &#x27;Variable&#x27;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">cox_df &lt;- rbind(colnames(cox_df),cox_df)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">forestplot(labeltext=as.matrix(cox_df)[,1:3],lower=as.numeric(cox_df[,4]),</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">           upper=as.numeric(cox_df[,5]),mean=as.numeric(cox_df[,6]),</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">           is.summary=c(T,F,F,F),</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">           zero=0,lineheight=unit(0.5,&#x27;cm&#x27;),graphwidth=unit(35,&#x27;mm&#x27;),</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">           colgap=unit(2,&#x27;mm&#x27;),lwd.zero=2,lwd.ci=3,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">           col=fpColors(box = &#x27;#7AC5CD&#x27;,lines = &#x27;black&#x27;,zero = &#x27;purple&#x27;),</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">           lwd.xaxis=2,lty.ci=&#x27;solid&#x27;,ci.vertices.height=0.05,graph.pos=2,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">           xlim=c(0,2.5),box.size=0.5,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">           hrzl_lines=list(&#x27;2&#x27;=gpar(lwd=3,columns=1:4,col=&#x27;dark green&#x27;),</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                           &#x27;5&#x27;=gpar(lwd=3,columns=1:4,col=&#x27;dark green&#x27;))</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">           )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain"># 多因素回归</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">res &lt;- coxph(Surv(OS.time,OS)~riskScore+gender+stage_event_pathologic_stage,data=df_forest)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">sum_res &lt;- summary(res)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">df &lt;- data.frame(row.names = rownames(sum_res$conf.int),pvalue=sum_res$coefficients[,&#x27;Pr(&gt;|z|)&#x27;],</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                 HR=sum_res$conf.int[,&#x27;exp(coef)&#x27;],lower=sum_res$conf.int[,&#x27;lower .95&#x27;],</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                 upper=sum_res$conf.int[,&#x27;upper .95&#x27;],means=res$means)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">df$pvalue &lt;- round(df$pvalue,3)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">df$HR &lt;- round(df$HR,3)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">df &lt;- rownames_to_column(df,var=&#x27;Variable&#x27;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">df &lt;- rbind(colnames(df),df)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">forestplot(labeltext=as.matrix(df)[,1:3],lower=as.numeric(df[,4]),</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">           upper=as.numeric(df[,5]),mean=as.numeric(df[,6]),</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">           is.summary=c(T,F,F,F),</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">           zero=0,lineheight=unit(0.5,&#x27;cm&#x27;),graphwidth=unit(35,&#x27;mm&#x27;),</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">           colgap=unit(2,&#x27;mm&#x27;),lwd.zero=2,lwd.ci=3,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">           col=fpColors(box = &#x27;#7AC5CD&#x27;,lines = &#x27;black&#x27;,zero = &#x27;purple&#x27;),</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">           lwd.xaxis=2,lty.ci=&#x27;solid&#x27;,ci.vertices.height=0.05,graph.pos=2,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">           xlim=c(0,2.5),box.size=0.5,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">           hrzl_lines=list(&#x27;2&#x27;=gpar(lwd=3,columns=1:4,col=&#x27;dark green&#x27;),</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                           &#x27;5&#x27;=gpar(lwd=3,columns=1:4,col=&#x27;dark green&#x27;))</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">)</span><br></span></code></pre><div class="buttonGroup__atx"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_eSgA" aria-hidden="true"><svg viewBox="0 0 24 24" class="copyButtonIcon_y97N"><path fill="currentColor" d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg viewBox="0 0 24 24" class="copyButtonSuccessIcon_LjdS"><path fill="currentColor" d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p><img loading="lazy" alt="image-20230509185354749" src="/BioinfoTech/assets/images/image-20230509185354749-588316ad581f51c25e2e7e25dea703f3.png" width="748" height="631" class="img_ev3q"></p><div class="language-R codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_biex"><pre tabindex="0" class="prism-code language-R codeBlock_bY9V thin-scrollbar"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token plain"># GSEA</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">df_exp &lt;- as.data.frame(t(exp))</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">df_exp &lt;- rownames_to_column(df_exp,var = &#x27;sample_ID&#x27;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">df_exp$sample_ID &lt;- substr(df_exp$sample_ID,1,12)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">df_exp &lt;- df_exp[!duplicated(df_exp$sample_ID),]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">rownames(df_exp) &lt;- df_exp$sample_ID</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">df_exp &lt;- as.data.frame(t(df_exp[,-1]))</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">df_exp &lt;- df_exp[,km_data$ID]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">group &lt;- as.factor(km_data$risk)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">design &lt;- model.matrix(~km_data$risk)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">fit &lt;- lmFit(df_exp,design)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">fit &lt;- eBayes(fit)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">df_deg &lt;- topTable(fit,coef = 2,number = Inf)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">df_deg$change &lt;- ifelse(df_deg$logFC&gt;log2(2) &amp; df_deg$adj.P.Val&lt;0.05,&#x27;up&#x27;,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                        ifelse(df_deg$logFC&lt;log2(0.5) &amp; df_deg$adj.P.Val&lt;0.05,&#x27;down&#x27;,&#x27;nochange&#x27;))</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">df_deg &lt;- df_deg[order(df_deg$logFC,decreasing = T),]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">symbol_2_entrez &lt;- bitr(rownames(df_deg),fromType = &#x27;SYMBOL&#x27;,toType = &#x27;ENTREZID&#x27;,OrgDb = org.Hs.eg.db)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">symbol_2_entrez &lt;- symbol_2_entrez[!duplicated(symbol_2_entrez$SYMBOL),]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">rownames(symbol_2_entrez) &lt;- symbol_2_entrez$SYMBOL</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">df_deg$entrezid &lt;- symbol_2_entrez[rownames(df_deg),&#x27;ENTREZID&#x27;]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">df_deg &lt;- df_deg[!duplicated(df_deg$entrezid),]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">df_deg &lt;- na.omit(df_deg)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">a &lt;- df_deg[(df_deg$change==&#x27;up&#x27;),&#x27;logFC&#x27;]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">names(a) &lt;- df_deg[(df_deg$change==&#x27;up&#x27;),&#x27;entrezid&#x27;]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">res_a &lt;- gseKEGG(a,&#x27;hsa&#x27;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">b &lt;- df_deg[(df_deg$change==&#x27;down&#x27;),&#x27;logFC&#x27;]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">names(b) &lt;- df_deg[(df_deg$change==&#x27;down&#x27;),&#x27;entrezid&#x27;]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">res_b &lt;- gseKEGG(b,&#x27;hsa&#x27;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">gseaplot2(res_a,geneSetID = 1:5)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">gseaplot2(res_b,geneSetID = 1:5)</span><br></span></code></pre><div class="buttonGroup__atx"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_eSgA" aria-hidden="true"><svg viewBox="0 0 24 24" class="copyButtonIcon_y97N"><path fill="currentColor" d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg viewBox="0 0 24 24" class="copyButtonSuccessIcon_LjdS"><path fill="currentColor" d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>先做差异分析，找到 high risk vs low risk 的差异基因，上调和下调的 gene 分别拿去做 GSEA 再画图</p><p><img loading="lazy" alt="image-20230509185420806" src="/BioinfoTech/assets/images/image-20230509185420806-5e537d74a87e1a2d5eec52b986a3fbef.png" width="748" height="528" class="img_ev3q"></p><div class="language-R codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_biex"><pre tabindex="0" class="prism-code language-R codeBlock_bY9V thin-scrollbar"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token plain"># waterfall plot</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">query &lt;- GDCquery(project = &#x27;TCGA-LUAD&#x27;,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                  data.category = &quot;Simple Nucleotide Variation&quot;,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                  data.type = &quot;Masked Somatic Mutation&quot;,</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">                  access = &#x27;open&#x27;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">GDCdownload(query)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">GDCprepare(query,save = T,save.filename = &#x27;TCGA-LUAD-SNP.Rdata&#x27;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">load(&#x27;TCGA-LUAD-SNP.Rdata&#x27;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">luad_snp &lt;- read.maf(maf=data)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">plotmafSummary(luad_snp,rmOutlier = TRUE, addStat = &#x27;median&#x27;, dashboard = TRUE, titvRaw = FALSE)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">oncoplot(luad_snp,top = 15)</span><br></span></code></pre><div class="buttonGroup__atx"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_eSgA" aria-hidden="true"><svg viewBox="0 0 24 24" class="copyButtonIcon_y97N"><path fill="currentColor" d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg viewBox="0 0 24 24" class="copyButtonSuccessIcon_LjdS"><path fill="currentColor" d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p><img loading="lazy" alt="image-20230509185458438" src="/BioinfoTech/assets/images/image-20230509185458438-b1d15f4de9cf21a70c4b45a30731c90c.png" width="748" height="1040" class="img_ev3q"></p><h2 class="anchor anchorWithStickyNavbar_LWe7" id="step-7-高低风险组间的免疫浸润">Step 7. 高低风险组间的免疫浸润<a href="#step-7-高低风险组间的免疫浸润" class="hash-link" aria-label="Direct link to Step 7. 高低风险组间的免疫浸润" title="Direct link to Step 7. 高低风险组间的免疫浸润">​</a></h2><div class="language-R codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_biex"><pre tabindex="0" class="prism-code language-R codeBlock_bY9V thin-scrollbar"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token plain">library(GSVA)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">library(GSEABase)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">library(ggplot2)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">library(tidyr)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">rm(list = ls())</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">#1. 获取 geneSets</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain"># 下载 28 种免疫细胞的参考基因集 &lt;http://cis.hku.hk/TISIDB/data/download/CellReports.txt&gt;</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">geneSet &lt;- read.csv(&quot;cellReport.txt&quot;,header = F,sep = &quot;\t&quot;,)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">geneSet[1:5,1:5]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">geneSet &lt;- t(tibble::column_to_rownames(geneSet,var = &#x27;V1&#x27;))</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">gene_list &lt;- list()</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">for (i in colnames(geneSet)){</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  gene_list[[i]] &lt;- geneSet[,i]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  gene_list[[i]] &lt;-gene_list[[i]][gene_list[[i]]!=&#x27;&#x27;] </span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">}</span><br></span></code></pre><div class="buttonGroup__atx"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_eSgA" aria-hidden="true"><svg viewBox="0 0 24 24" class="copyButtonIcon_y97N"><path fill="currentColor" d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg viewBox="0 0 24 24" class="copyButtonSuccessIcon_LjdS"><path fill="currentColor" d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>直接读取的 geneSet 长这样，有空值，需要转变成 gene<!-- -->_<!-- -->list，list 长度为 28，其中每个元素是某个细胞所有基因的向量，名字是对应的细胞</p><p><img loading="lazy" alt="image-20230509185550240" src="/BioinfoTech/assets/images/image-20230509185550240-5e29866f751745e4d77bf5d0a2cfd934.png" width="743" height="471" class="img_ev3q"></p><div class="language-R codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_biex"><pre tabindex="0" class="prism-code language-R codeBlock_bY9V thin-scrollbar"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token plain"># 载入表达矩阵，把表达矩阵分成 high risk 和 low risk 组</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">load(&#x27;TCGA_LUAD_TPM_exp.Rdata&#x27;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">load(&#x27;risk.Rdata&#x27;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain"># 对表达矩阵取 log2</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">exp &lt;- log2(exp+0.00001)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">exp &lt;- as.data.frame(exp)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">colnames(exp) &lt;- substr(colnames(exp),1,12)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">exp &lt;- exp[,!duplicated(colnames(exp))]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">high_risk_sample &lt;- km_data[km_data$risk==&#x27;high&#x27;,&#x27;ID&#x27;]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">low_risk_sample &lt;- km_data[km_data$risk==&#x27;low&#x27;,&#x27;ID&#x27;]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">high_risk_exp &lt;- exp[,high_risk_sample]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">low_risk_exp &lt;- exp[,low_risk_sample]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain"># ssGSEA</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">high_risk_gsea &lt;- gsva(high_risk_exp,gene_list,method=&#x27;ssgsea&#x27;,kcdf=&#x27;Gaussian&#x27;,abs.ranking=T)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">low_risk_gsea &lt;- gsva(low_risk_exp,gene_list,method=&#x27;ssgsea&#x27;,kcdf=&#x27;Gaussian&#x27;,abs.ranking=T)</span><br></span></code></pre><div class="buttonGroup__atx"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_eSgA" aria-hidden="true"><svg viewBox="0 0 24 24" class="copyButtonIcon_y97N"><path fill="currentColor" d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg viewBox="0 0 24 24" class="copyButtonSuccessIcon_LjdS"><path fill="currentColor" d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>要注意 gsva 函数的输入得是 matrix，不能是 data.frame，不然会报错</p><div class="language-R codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_biex"><pre tabindex="0" class="prism-code language-R codeBlock_bY9V thin-scrollbar"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token plain">hr_gsea_norm &lt;- high_risk_gsea</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">for (i in colnames(hr_gsea_norm)) {</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  diff &lt;- max(hr_gsea_norm[,i] )-min(hr_gsea_norm[,i] )</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  hr_gsea_norm[,i] &lt;- (hr_gsea_norm[,i] -min(hr_gsea_norm[,i]))/diff</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">}</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">hr_gsea_norm &lt;- as.data.frame(t(hr_gsea_norm))</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">hr_gsea_norm$type &lt;- rep(&#x27;high&#x27;,times=nrow(hr_gsea_norm))</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">lr_gsea_norm &lt;- low_risk_gsea</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">for (i in colnames(lr_gsea_norm)) {</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  diff &lt;- max(lr_gsea_norm[,i] )-min(lr_gsea_norm[,i])</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  lr_gsea_norm[,i] &lt;- (lr_gsea_norm[,i] -min(lr_gsea_norm[,i]))/diff</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">}</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">lr_gsea_norm &lt;- as.data.frame(t(lr_gsea_norm))</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">lr_gsea_norm$type &lt;- rep(&#x27;low&#x27;,times=nrow(lr_gsea_norm))</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">gsea_norm &lt;- rbind(hr_gsea_norm,lr_gsea_norm)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">longer &lt;- pivot_longer(gsea_norm,cols = !type,names_to = &#x27;cell_type&#x27;,values_to = &#x27;Score&#x27;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">ggplot(longer,aes(x=cell_type,y=Score,color=type))+geom_boxplot()+</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  theme(axis.text.x = element_text(angle = 45,hjust = 1))+</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  stat_compare_means(label = &quot;p.signif&quot;,size=3,method = &#x27;wilcox.test&#x27;)</span><br></span></code></pre><div class="buttonGroup__atx"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_eSgA" aria-hidden="true"><svg viewBox="0 0 24 24" class="copyButtonIcon_y97N"><path fill="currentColor" d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg viewBox="0 0 24 24" class="copyButtonSuccessIcon_LjdS"><path fill="currentColor" d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>先把 high risk gsea 和 low risk gsea 的数据分别进行 min max scale，新增一列标记上 high 和 low，然后合并起来，行是样本，列是细胞。再把透视表转成长表，就可以画图了</p><p><img loading="lazy" alt="image-20230509185611592" src="/BioinfoTech/assets/images/image-20230509185611592-c6ea283be60e70e9dd77be1fc0e5a80b.png" width="743" height="531" class="img_ev3q"></p><div class="language-R codeBlockContainer_Ckt0 theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_biex"><pre tabindex="0" class="prism-code language-R codeBlock_bY9V thin-scrollbar"><code class="codeBlockLines_e6Vv"><span class="token-line" style="color:#393A34"><span class="token plain"># 免疫检查点&amp;TBM</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">risk_exp &lt;- as.data.frame(t(exp))</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">risk_exp &lt;- risk_exp[km_data$ID,]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">risk_exp$type &lt;- km_data$risk</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">risk_exp &lt;- risk_exp[,c(&#x27;CD274&#x27;,&#x27;PDCD1&#x27;,&#x27;CTLA4&#x27;,&#x27;LAG3&#x27;,&#x27;TIGIT&#x27;,&#x27;type&#x27;)]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">risk_long &lt;- pivot_longer(risk_exp,cols = !type,names_to = &#x27;ICIs&#x27;,values_to = &#x27;Expression&#x27;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">ggplot(risk_long,aes(x=ICIs,y=Expression,fill=type))+geom_boxplot()+</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">  stat_compare_means(label = &quot;p.format&quot;,size=3,method = &#x27;wilcox.test&#x27;)</span><br></span></code></pre><div class="buttonGroup__atx"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_eSgA" aria-hidden="true"><svg viewBox="0 0 24 24" class="copyButtonIcon_y97N"><path fill="currentColor" d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg viewBox="0 0 24 24" class="copyButtonSuccessIcon_LjdS"><path fill="currentColor" d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p><img loading="lazy" alt="image-20230509185633127" src="/BioinfoTech/assets/images/image-20230509185633127-d63daaf6acfe39a78b132b3b734f99df.png" width="743" height="527" class="img_ev3q"></p></div><footer class="theme-doc-footer docusaurus-mt-lg"><div class="theme-doc-footer-edit-meta-row row"><div class="col"><a href="https://github.com/facebook/docusaurus/tree/main/packages/create-docusaurus/templates/shared/docs/03-常用分析流程及教程/复现.md" target="_blank" rel="noreferrer noopener" class="theme-edit-this-page"><svg fill="currentColor" height="20" width="20" viewBox="0 0 40 40" class="iconEdit_Z9Sw" aria-hidden="true"><g><path d="m34.5 11.7l-3 3.1-6.3-6.3 3.1-3q0.5-0.5 1.2-0.5t1.1 0.5l3.9 3.9q0.5 0.4 0.5 1.1t-0.5 1.2z m-29.5 17.1l18.4-18.5 6.3 6.3-18.4 18.4h-6.3v-6.2z"></path></g></svg>Edit this page</a></div><div class="col lastUpdated_vwxv"></div></div></footer></article><nav class="pagination-nav docusaurus-mt-lg" aria-label="Docs pages"><a class="pagination-nav__link pagination-nav__link--prev" href="/BioinfoTech/软件安装及环境配置/省医服务器的使用"><div class="pagination-nav__sublabel">Previous</div><div class="pagination-nav__label">省医服务器的使用</div></a><a class="pagination-nav__link pagination-nav__link--next" href="/BioinfoTech/常用分析流程及教程/拟时序分析"><div class="pagination-nav__sublabel">Next</div><div class="pagination-nav__label">单细胞数据（loom）的拟时序分析</div></a></nav></div></div><div class="col col--3"><div class="tableOfContents_bqdL thin-scrollbar theme-doc-toc-desktop"><ul class="table-of-contents table-of-contents__left-border"><li><a href="#step-1-单细胞数据处理" class="table-of-contents__link toc-highlight">Step 1. 单细胞数据处理</a><ul><li><a href="#11-qc" class="table-of-contents__link toc-highlight">1.1 QC</a></li><li><a href="#12-用-harmony-去除批次效应" class="table-of-contents__link toc-highlight">1.2 用 harmony 去除批次效应</a></li><li><a href="#13-细胞注释" class="table-of-contents__link toc-highlight">1.3 细胞注释</a></li><li><a href="#14-鉴定关键细胞类型" class="table-of-contents__link toc-highlight">1.4 鉴定关键细胞类型</a></li><li><a href="#15-单细胞通路富集" class="table-of-contents__link toc-highlight">1.5 单细胞通路富集</a></li><li><a href="#16-拟时序分析" class="table-of-contents__link toc-highlight">1.6 拟时序分析</a></li><li><a href="#17-细胞通讯" class="table-of-contents__link toc-highlight">1.7 细胞通讯</a></li></ul></li><li><a href="#step-2-tcga-差异分析" class="table-of-contents__link toc-highlight">Step 2. TCGA 差异分析</a></li><li><a href="#step-3-wgcna-分析" class="table-of-contents__link toc-highlight">Step 3. WGCNA 分析</a></li><li><a href="#step-4-非负矩阵分解算法" class="table-of-contents__link toc-highlight">Step 4. 非负矩阵分解算法</a><ul><li><a href="#41-nmf-算法对样本分类" class="table-of-contents__link toc-highlight">4.1 NMF 算法对样本分类</a></li><li><a href="#42-生存分析" class="table-of-contents__link toc-highlight">4.2 生存分析</a></li><li><a href="#43-免疫浸润" class="table-of-contents__link toc-highlight">4.3 免疫浸润</a></li><li><a href="#44-批量单因素-cox-回归" class="table-of-contents__link toc-highlight">4.4 批量单因素 Cox 回归</a></li></ul></li><li><a href="#step-5-预后模型构建和验证" class="table-of-contents__link toc-highlight">Step 5. 预后模型构建和验证</a></li><li><a href="#step-6-riskscore-和临床信息的相关性" class="table-of-contents__link toc-highlight">Step 6. riskScore 和临床信息的相关性</a></li><li><a href="#step-7-高低风险组间的免疫浸润" class="table-of-contents__link toc-highlight">Step 7. 高低风险组间的免疫浸润</a></li></ul></div></div></div></div></main></div></div><footer class="footer footer--dark"><div class="container container-fluid"><div class="row footer__links"><div class="col footer__col"><div class="footer__title">Docs</div><ul class="footer__items clean-list"><li class="footer__item"><a class="footer__link-item" href="/BioinfoTech/intro">Tutorial</a></li></ul></div><div class="col footer__col"><div class="footer__title">Community</div><ul class="footer__items clean-list"><li class="footer__item"><a href="https://stackoverflow.com/questions/tagged/docusaurus" target="_blank" rel="noopener noreferrer" class="footer__link-item">Stack Overflow<svg width="13.5" height="13.5" aria-hidden="true" viewBox="0 0 24 24" class="iconExternalLink_nPIU"><path fill="currentColor" d="M21 13v10h-21v-19h12v2h-10v15h17v-8h2zm3-12h-10.988l4.035 4-6.977 7.07 2.828 2.828 6.977-7.07 4.125 4.172v-11z"></path></svg></a></li><li class="footer__item"><a href="https://discordapp.com/invite/docusaurus" target="_blank" rel="noopener noreferrer" class="footer__link-item">Discord<svg width="13.5" height="13.5" aria-hidden="true" viewBox="0 0 24 24" class="iconExternalLink_nPIU"><path fill="currentColor" d="M21 13v10h-21v-19h12v2h-10v15h17v-8h2zm3-12h-10.988l4.035 4-6.977 7.07 2.828 2.828 6.977-7.07 4.125 4.172v-11z"></path></svg></a></li><li class="footer__item"><a href="https://twitter.com/docusaurus" target="_blank" rel="noopener noreferrer" class="footer__link-item">Twitter<svg width="13.5" height="13.5" aria-hidden="true" viewBox="0 0 24 24" class="iconExternalLink_nPIU"><path fill="currentColor" d="M21 13v10h-21v-19h12v2h-10v15h17v-8h2zm3-12h-10.988l4.035 4-6.977 7.07 2.828 2.828 6.977-7.07 4.125 4.172v-11z"></path></svg></a></li></ul></div><div class="col footer__col"><div class="footer__title">More</div><ul class="footer__items clean-list"><li class="footer__item"><a href="https://github.com/WhyLIM/BioinfoTech" target="_blank" rel="noopener noreferrer" class="footer__link-item">GitHub<svg width="13.5" height="13.5" aria-hidden="true" viewBox="0 0 24 24" class="iconExternalLink_nPIU"><path fill="currentColor" d="M21 13v10h-21v-19h12v2h-10v15h17v-8h2zm3-12h-10.988l4.035 4-6.977 7.07 2.828 2.828 6.977-7.07 4.125 4.172v-11z"></path></svg></a></li></ul></div></div><div class="footer__bottom text--center"><div class="footer__copyright">Copyright © 2023 Lab TD, Inc. Built with Docusaurus.</div></div></div></footer></div>
<script src="/BioinfoTech/assets/js/runtime~main.9748e4a1.js"></script>
<script src="/BioinfoTech/assets/js/main.3ef7b66f.js"></script>
</body>
</html>